// @platform-ui/inputTag/esm/Droplist/Droplist.js
import React from 'react';
import { DropdownList } from '@platform-ui/dropdown';
import { useTagsContextAutocomplete } from '../TagsContext';

const Droplist = ({
  children
}) => {
  const context = useTagsContextAutocomplete();
  const {
    size,
    opened,
    loading,
    noResults,
    options,
    editing,
    setEditing,
    focusedIndex,
    setInputValue,
    highlightedOption,
    setHighlightedOption,
    onAdd,
    onReplace,
    dataQaType,
    acceptTag
  } = context;
  return /*#__PURE__*/React.createElement(DropdownList, {
    dataQaType: `${dataQaType}.droplist`,
    size: size,
    opened: opened,
    options: options,
    maxHeight: loading || noResults ? 'auto' : undefined,
    highlightedOption: highlightedOption || undefined,
    onOptionClick: (event, params) => {
      const tag = acceptTag(params.option);

      if (tag === null) {
        return;
      }

      setHighlightedOption(null);

      if (editing) {
        setEditing(false);
        onReplace(event, {
          index: focusedIndex,
          tag
        });
      } else {
        onAdd(event, {
          tag
        });
      }
    },
    onOptionMouseOver: (event, params) => {
      setHighlightedOption(params.option);
    },
    onMouseLeave: () => {
      setHighlightedOption(null);
    }
  }, children);
};

export default Droplist;
