// @platform-ui/tooltip/TooltipBubble/TooltipBubble.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.props = exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _TooltipBubbleModule = _interopRequireDefault(require("./TooltipBubble.module.css"));

var _excluded = ["alignText", "children", "cornerStyle", "direction", "inline", "withCorner", "theme", "id", "overflow", "dataQaType"];
var props = {
  theme: ['dark', 'light'],
  align: ['left', 'center', 'right', 'inherit', 'justify']
};
exports.props = props;

var TooltipBubble = function TooltipBubble(_ref) {
  var _classNames;

  var alignText = _ref.alignText,
      children = _ref.children,
      cornerStyle = _ref.cornerStyle,
      direction = _ref.direction,
      _ref$inline = _ref.inline,
      inline = _ref$inline === void 0 ? false : _ref$inline,
      _ref$withCorner = _ref.withCorner,
      withCorner = _ref$withCorner === void 0 ? true : _ref$withCorner,
      _ref$theme = _ref.theme,
      theme = _ref$theme === void 0 ? 'light' : _ref$theme,
      id = _ref.id,
      _ref$overflow = _ref.overflow,
      overflow = _ref$overflow === void 0 ? 'hidden' : _ref$overflow,
      dataQaType = _ref.dataQaType,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", Object.assign({
    className: _TooltipBubbleModule.default.bubbleWrapper,
    id: id
  }, (0, _dataQa.default)(otherProps)), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      overflow: overflow
    },
    className: (0, _classnames.default)((_classNames = {}, _classNames[_TooltipBubbleModule.default.bubble] = true, _classNames[_TooltipBubbleModule.default.bubble_inline] = inline, _classNames[_TooltipBubbleModule.default["bubble_direction_" + direction]] = direction, _classNames[_TooltipBubbleModule.default["bubble_theme_" + theme]] = theme, _classNames))
  }, withCorner && /*#__PURE__*/_react.default.createElement("div", {
    style: cornerStyle,
    className: _TooltipBubbleModule.default.corner
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _TooltipBubbleModule.default.bubbleContent,
    style: {
      textAlign: alignText
    },
    "data-qa-type": dataQaType + ".content"
  }, children)));
};

var _default = TooltipBubble;
exports.default = _default;
