// @platform-ui/inputTag/TagsList/TagsList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _scrollBlock = _interopRequireDefault(require("@platform-ui/scrollBlock"));

var _TagsListModule = _interopRequireDefault(require("./TagsList.module.css"));

var TagsList = function TagsList(_ref) {
  var focused = _ref.focused,
      children = _ref.children,
      scrollable = _ref.scrollable,
      size = _ref.size;
  var setScrollPosition = (0, _react.useRef)();

  var renderList = function renderList() {
    var childrenArray = _react.default.Children.toArray(children);

    return childrenArray.map(function (child, index) {
      return /*#__PURE__*/_react.default.createElement("span", {
        key: index === childrenArray.length - 1 ? 'input' : index,
        className: _TagsListModule.default.item,
        onMouseDown: function onMouseDown(event) {
          event.stopPropagation();
        }
      }, child);
    });
  };

  var getScrollBlockControllers = function getScrollBlockControllers(_ref2) {
    var fn = _ref2.setScrollPosition;
    setScrollPosition.current = fn;
  };

  (0, _react.useEffect)(function () {
    if (focused) {
      setScrollPosition.current == null ? void 0 : setScrollPosition.current('right');
    } else {
      setScrollPosition.current == null ? void 0 : setScrollPosition.current('left');
    }
  }, [focused]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_TagsListModule.default.list, _TagsListModule.default["list_size_" + size])
  }, scrollable ? /*#__PURE__*/_react.default.createElement(_scrollBlock.default, {
    getControllers: getScrollBlockControllers
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _TagsListModule.default.row
  }, renderList())) : renderList());
};

var _default = TagsList;
exports.default = _default;
