// @platform-ui/label/withLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.withLabel = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _Label = _interopRequireDefault(require("./Label"));

var _LabelModule = _interopRequireDefault(require("./Label.module.css"));

var _excluded = ["dataQaType", "dataQaData", "children", "size", "disabled", "hintBlock"];

var withLabel = function withLabel(Component, defaults) {
  if (defaults === void 0) {
    defaults = {};
  }

  return function (props) {
    var _props$dataQaType = props.dataQaType,
        dataQaType = _props$dataQaType === void 0 ? defaults.dataQaType : _props$dataQaType,
        dataQaData = props.dataQaData,
        children = props.children,
        _props$size = props.size,
        size = _props$size === void 0 ? 'l' : _props$size,
        disabled = props.disabled,
        hintBlock = props.hintBlock,
        inputProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

    var input = /*#__PURE__*/_react.default.createElement(Component, Object.assign({}, inputProps, {
      size: size,
      disabled: disabled,
      dataQaType: dataQaType + ".input"
    }));

    return /*#__PURE__*/_react.default.createElement(_Label.default, {
      dataQaType: dataQaType,
      dataQaData: dataQaData,
      size: size,
      input: input,
      disabled: disabled
    }, children, hintBlock ? /*#__PURE__*/_react.default.createElement("span", {
      className: _LabelModule.default.hintBlock
    }, hintBlock) : null);
  };
};

exports.withLabel = withLabel;
