// @platform-ui/navigation/Link/Link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _excluded = ["external", "onClick", "path", "dataQaType", "dataQaData"];

var handleLinkClick = function handleLinkClick(_ref) {
  var path = _ref.path,
      onClick = _ref.onClick;
  return function (e) {
    onClick && onClick(e, {
      path: path
    });
  };
};

var Link = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var external = props.external,
      onClick = props.onClick,
      path = props.path,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  var handleMouseDown = function handleMouseDown(e) {
    e.preventDefault();
  };

  var handleClick = function handleClick(e) {
    return handleLinkClick(props)(e);
  };

  return /*#__PURE__*/_react.default.createElement("a", Object.assign({
    ref: ref,
    href: path,
    onMouseDown: handleMouseDown,
    onClick: handleClick
  }, restProps, (0, _dataQa.default)(props)));
});
var _default = Link;
exports.default = _default;
