// @platform-ui/notification/esm/NotificationStack/NotificationStack.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["dataQaType", "paddingBottom", "notifications", "offsetTop", "offsetRight", "width", "zIndex", "onHide", "onClose", "onActionConfirm", "onCancel"],
      _excluded2 = ["notificationComponent"];
import React from 'react';
import Notification from '../NotificationToast/NotificationToast';
import styles from './NotificationStack.module.css';

const NotificationStack = props => {
  const {
    dataQaType = 'uikit/NotificationStack',
    // TODO: переименовать проп так, чтобы было понятно что именно он делает - задаёт отступ между нотификациями
    paddingBottom = 16,
    notifications,
    offsetTop,
    offsetRight,
    width,
    zIndex,
    onHide,
    onClose,
    onActionConfirm,
    onCancel
  } = props,
        baseProps = _objectWithoutPropertiesLoose(props, _excluded);

  return /*#__PURE__*/React.createElement("div", {
    "data-qa-type": dataQaType,
    className: styles.root,
    style: {
      top: offsetTop,
      right: offsetRight,
      width,
      zIndex
    }
  }, Array.isArray(notifications) && notifications.map((notification, index) => {
    const {
      notificationComponent: NotificationComponent = Notification
    } = notification,
          notificationProps = _objectWithoutPropertiesLoose(notification, _excluded2);

    return /*#__PURE__*/React.createElement(NotificationComponent, Object.assign({
      key: notificationProps.id || index,
      onHide: onHide ? event => onHide(event, notificationProps) : undefined,
      onClose: onClose ? event => onClose(event, notificationProps) : undefined,
      onActionConfirm: onActionConfirm ? event => onActionConfirm(event, notificationProps) : undefined,
      onCancel: onCancel ? event => onCancel(event, notificationProps) : undefined,
      paddingBottom: paddingBottom
    }, baseProps, notificationProps));
  }));
}; // TODO: убрать в мажорном обновлении


NotificationStack.defaultProps = _extends({}, Notification.defaultProps, {
  dataQaType: 'uikit/NotificationStack',
  // TODO: переименовать проп так, чтобы было понятно что именно он делает - задаёт отступ между нотификациями
  paddingBottom: 16
});
export default NotificationStack;
