// @platform-ui/tooltip/esm/TooltipBubble/TooltipBubble.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["alignText", "children", "cornerStyle", "direction", "inline", "withCorner", "theme", "id", "overflow", "dataQaType"];
import React from 'react';
import classNames from 'classnames';
import dataQa from '@platform-ui/dataQa';
import styles from './TooltipBubble.module.css';
export const props = {
  theme: ['dark', 'light'],
  align: ['left', 'center', 'right', 'inherit', 'justify']
};

const TooltipBubble = _ref => {
  let {
    alignText,
    children,
    cornerStyle,
    direction,
    inline = false,
    withCorner = true,
    theme = 'light',
    id,
    overflow = 'hidden',
    dataQaType
  } = _ref,
      otherProps = _objectWithoutPropertiesLoose(_ref, _excluded);

  return /*#__PURE__*/React.createElement("div", Object.assign({
    className: styles.bubbleWrapper,
    id: id
  }, dataQa(otherProps)), /*#__PURE__*/React.createElement("div", {
    style: {
      overflow
    },
    className: classNames({
      [styles.bubble]: true,
      [styles.bubble_inline]: inline,
      [styles[`bubble_direction_${direction}`]]: direction,
      [styles[`bubble_theme_${theme}`]]: theme
    })
  }, withCorner && /*#__PURE__*/React.createElement("div", {
    style: cornerStyle,
    className: styles.corner
  }), /*#__PURE__*/React.createElement("div", {
    className: styles.bubbleContent,
    style: {
      textAlign: alignText
    },
    "data-qa-type": `${dataQaType}.content`
  }, children)));
};

export default TooltipBubble;
