// @platform-ui/link/Link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = exports.THEMES = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _clickable = _interopRequireDefault(require("@platform-ui/clickable"));

var _LinkModule = _interopRequireDefault(require("./Link.module.css"));

var _excluded = ["dataQaType", "pseudo", "theme"];
var THEMES = ['default', 'light', 'dark'];
exports.THEMES = THEMES;
var Link = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _cn;

  var dataQaType = _ref.dataQaType,
      pseudo = _ref.pseudo,
      theme = _ref.theme,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var themeType = typeof theme !== 'object' ? theme : 'custom';
  var customTheme = theme;
  var style = {
    '--tds-component--link--color--default': customTheme.color || 'var(--tds-color-link)',
    '--tds-component--link--color--hover': customTheme.colorHover || customTheme.color || 'var(--tds-color-link-hover)',
    '--tds-component--link--color--focus': customTheme.colorFocus || customTheme.color || 'var(--tds-color-link)',
    '--tds-component--link--background-color--default': customTheme.backgroundColor || 'transparent',
    '--tds-component--link--background-color--hover': customTheme.backgroundColorHover || 'transparent',
    '--tds-component--link--background-color--focus': customTheme.backgroundColorFocus || 'var(--tds-color-warning-bg)'
  };
  return /*#__PURE__*/_react.default.createElement(_clickable.default, Object.assign({}, otherProps, {
    dataQaType: dataQaType,
    className: (0, _classnames.default)(_LinkModule.default.link, (_cn = {}, _cn[_LinkModule.default["link_theme_" + themeType]] = themeType, _cn[_LinkModule.default.link_pseudo] = pseudo, _cn[_LinkModule.default.link_disabled] = otherProps.disabled, _cn)),
    innerClassName: _LinkModule.default.inner,
    innerDataQaType: dataQaType + ".inner",
    style: typeof theme === 'object' ? style : {},
    ref: ref
  }));
});
Link.defaultProps = {
  dataQaType: 'uikit/link',
  theme: 'default'
};
Link.displayName = 'Link';
var _default = Link;
exports.default = _default;
