// @platform-ui/badge/Badge.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var corners = _interopRequireWildcard(require("@platform-ui/flatCorners"));

var _clickable = _interopRequireDefault(require("@platform-ui/clickable"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _color = require("@platform-ui/utils/color");

var _getBackgroundMainColor = _interopRequireDefault(require("@platform-ui/utils/getBackgroundMainColor"));

var _BadgeModule = _interopRequireDefault(require("./Badge.module.css"));

var _excluded = ["borderRadius", "children", "color", "customBackground", "customColor", "hover", "padding", "size", "status", "statusColor", "statusCustomColor", "flatCorners", "theme", "dataQaType", "href", "tagName"];

var Badge = function Badge(_ref) {
  var _cn, _cn2;

  var borderRadius = _ref.borderRadius,
      children = _ref.children,
      color = _ref.color,
      customBackground = _ref.customBackground,
      customColor = _ref.customColor,
      hover = _ref.hover,
      padding = _ref.padding,
      size = _ref.size,
      status = _ref.status,
      statusColor = _ref.statusColor,
      statusCustomColor = _ref.statusCustomColor,
      flatCorners = _ref.flatCorners,
      theme = _ref.theme,
      dataQaType = _ref.dataQaType,
      href = _ref.href,
      tagName = _ref.tagName,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var smartColor = customColor || customBackground && (0, _color.calcFontColor)((0, _getBackgroundMainColor.default)(customBackground));
  return !!children && /*#__PURE__*/_react.default.createElement(_clickable.default, Object.assign({}, otherProps, {
    dataQaType: dataQaType,
    className: (0, _classnames.default)((_cn = {}, _cn[_BadgeModule.default.root] = true, _cn[_BadgeModule.default.root_hover] = hover, _cn[_BadgeModule.default["root_size_" + size]] = size, _cn[_BadgeModule.default["root_color_" + color]] = color && !theme, _cn[_BadgeModule.default["root_theme_" + theme]] = theme, _cn[_BadgeModule.default.root_color_custom] = customBackground, _cn)),
    tagName: typeof href === 'string' ? 'a' : tagName,
    href: href,
    style: (0, _extends2.default)({
      color: smartColor,
      borderRadius: borderRadius,
      background: customBackground && customBackground + " border-box"
    }, corners.style(flatCorners)),
    innerDataQaType: dataQaType + ".innerWrapper",
    innerClassName: _BadgeModule.default.innerClickable
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: _BadgeModule.default.innerClassname,
    style: {
      paddingRight: padding && +padding,
      paddingLeft: padding && +padding
    }
  }, status && /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".status"
  }), {
    className: (0, _classnames.default)((_cn2 = {}, _cn2[_BadgeModule.default.status] = true, _cn2[_BadgeModule.default["status_color_" + statusColor]] = statusColor && !statusCustomColor, _cn2[_BadgeModule.default["status_theme_" + theme]] = theme && !statusColor && !statusCustomColor, _cn2)),
    style: {
      backgroundColor: statusCustomColor
    }
  })), /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".content"
  }), {
    className: _BadgeModule.default.content
  }), children))) || null;
};

Badge.defaultProps = {
  dataQaType: 'uikit/badge',
  color: 'gray',
  padding: 12,
  size: 'l',
  tabIndex: -1,
  tagName: 'span'
};
var _default = Badge;
exports.default = _default;
