// @platform-ui/constants/mediaBreakpoints.js
"use strict";

var bp = require('./media');

module.exports = {
  phone: "(max-width: " + bp.PHONE + "px)",
  phoneSmall: "(max-width: " + bp.PHONE_SMALL + "px)",
  phoneGigant: "(min-width: " + bp.PHONE_GIANT + "px) and (max-width: " + bp.PHONE + "px)",
  tablets: "(min-width: " + bp.TABLET_S + "px)",
  onlyTablets: "(min-width: " + bp.TABLET_S + "px) and (max-width: " + bp.MOBILE + "px)",
  mobile: "(max-width: " + bp.MOBILE + "px)",
  desktop: "(min-width: " + bp.DESKTOP_S + "px)",
  desktopSmall: "(min-width: " + bp.DESKTOP_M + "px)",
  desktopMedium: "(min-width: " + bp.DESKTOP_M + "px)",
  desktopLarge: "(min-width: " + bp.DESKTOP_L + "px)",
  desktopDambo: "(min-width: " + bp.DESKTOP_L + "px)",
  desktopExtraLarge: "(min-width: " + bp.DESKTOP_EXTRA_LARGE + "px)",
  smallHeight: '(max-height: 480px)',
  retina: '(-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi)',
  largeBoardOnDesktop: "(min-width: " + bp.DESKTOP_L + "px)",
  // Новые брейкпоинты
  tabletS: "(min-width: " + bp.TABLET_S + "px)",
  tabletL: "(min-width: " + bp.TABLET_L + "px)",
  desktopS: "(min-width: " + bp.DESKTOP_S + "px)",
  desktopM: "(min-width: " + bp.DESKTOP_M + "px)",
  desktopMd: "(min-width: " + bp.DESKTOP_MD + "px)",
  desktopL: "(min-width: " + bp.DESKTOP_L + "px)"
};
