// @platform-ui/tooltip/Hint/Hint.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Tooltip = _interopRequireDefault(require("../Tooltip"));

var _icons = require("../icons");

var _HintModule = _interopRequireDefault(require("./Hint.module.css"));

var _excluded = ["hintColor", "focusable"];

var Hint = function Hint(_ref) {
  var _cn;

  var _ref$hintColor = _ref.hintColor,
      hintColor = _ref$hintColor === void 0 ? 'greyLighten20' : _ref$hintColor,
      focusable = _ref.focusable,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);

  var _React$useState = _react.default.useState(false),
      opened = _React$useState[0],
      setOpened = _React$useState[1];

  var handleFocus = function handleFocus() {
    setOpened(true);
  };

  var handleBlur = function handleBlur() {
    setOpened(false);
  };

  var tooltipProps = focusable ? {
    isActive: otherProps.isActive || opened,
    isInactive: otherProps.isInactive || !opened,
    onShow: handleFocus,
    onClose: handleBlur
  } : {};
  var wrapperProps = focusable ? {
    tabIndex: 0,
    onFocus: handleFocus,
    onBlur: handleBlur
  } : {};
  return /*#__PURE__*/_react.default.createElement(_Tooltip.default, Object.assign({}, otherProps, tooltipProps), /*#__PURE__*/_react.default.createElement("span", Object.assign({}, wrapperProps, {
    className: (0, _classnames.default)((_cn = {}, _cn[_HintModule.default.wrapper] = true, _cn[_HintModule.default["wrapper_color_" + hintColor]] = hintColor, _cn))
  }), /*#__PURE__*/_react.default.createElement(_icons.Hint, null)));
};

var _default = Hint;
exports.default = _default;
