// @platform-ui/styleProps/padding/index.js
"use strict";

exports.__esModule = true;
exports.usePaddingStyle = exports.usePaddingProps = exports.pickPaddingProps = exports.calcPaddingStyle = void 0;

var _react = require("react");

var _utils = require("../utils");

var pickPaddingProps = (0, _utils.createPropsPicker)(['padding', 'paddingX', 'paddingY', 'paddingTop', 'paddingLeft', 'paddingRight', 'paddingBottom']);
exports.pickPaddingProps = pickPaddingProps;

var calcPaddingStyle = function calcPaddingStyle(props) {
  var p = props.padding,
      pX = props.paddingX,
      pY = props.paddingY,
      pTop = props.paddingTop,
      pLeft = props.paddingLeft,
      pRight = props.paddingRight,
      pBottom = props.paddingBottom;
  var pA = (0, _utils.splitShorthandProperty)(p);
  var paddingTop = (0, _utils.pickFirstDefined)(pTop, pY, pA[0]);
  var paddingLeft = (0, _utils.pickFirstDefined)(pLeft, pX, pA[3], pA[1], pA[0]);
  var paddingRight = (0, _utils.pickFirstDefined)(pRight, pX, pA[1], pA[0]);
  var paddingBottom = (0, _utils.pickFirstDefined)(pBottom, pY, pA[2], pA[0]);
  return {
    paddingTop: paddingTop,
    paddingLeft: paddingLeft,
    paddingRight: paddingRight,
    paddingBottom: paddingBottom
  };
};

exports.calcPaddingStyle = calcPaddingStyle;

var usePaddingStyle = function usePaddingStyle(props) {
  var padding = props.padding,
      paddingX = props.paddingX,
      paddingY = props.paddingY,
      paddingTop = props.paddingTop,
      paddingLeft = props.paddingLeft,
      paddingRight = props.paddingRight,
      paddingBottom = props.paddingBottom;
  return (0, _react.useMemo)(function () {
    return calcPaddingStyle({
      padding: padding,
      paddingX: paddingX,
      paddingY: paddingY,
      paddingTop: paddingTop,
      paddingLeft: paddingLeft,
      paddingRight: paddingRight,
      paddingBottom: paddingBottom
    });
  }, [padding, paddingX, paddingY, paddingTop, paddingLeft, paddingRight, paddingBottom]);
};

exports.usePaddingStyle = usePaddingStyle;
var usePaddingProps = (0, _utils.createPropsHook)(pickPaddingProps, usePaddingStyle);
exports.usePaddingProps = usePaddingProps;
