// @platform-ui/navigation/esm/MobileMenu/MobileMenuItem.js
import React, { forwardRef } from 'react';
import cn from 'classnames';
import styles from './MobileMenuItem.module.css';
import MenuItemWithDropdown from '../MenuItemWithDropdown';
import Link from '../Link';
import Badge, { NewSection } from '../Badge';
import { useNavigationContext } from '../NavigationContext';
const MobileMenuItem = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    title,
    active,
    underlined,
    path,
    icon,
    count,
    onClick,
    dataQaType,
    dataQaData,
    newItem,
    dropdownItems
  } = props;
  const {
    theme
  } = useNavigationContext();
  return /*#__PURE__*/React.createElement("span", {
    className: cn(styles.root, {
      [styles[`root_theme_${theme}`]]: theme
    })
  }, /*#__PURE__*/React.createElement(MenuItemWithDropdown, {
    dropdownItems: dropdownItems,
    onClick: onClick
  }, /*#__PURE__*/React.createElement(Link, {
    ref: ref,
    className: cn(styles.link, {
      [styles.link_active]: active,
      [styles.link_underlined]: underlined
    }),
    "aria-current": active ? 'page' : undefined,
    path: path,
    onClick: onClick,
    dataQaType: dataQaType || 'uikit/navigation.mobileMenuItem',
    dataQaData: dataQaData
  }, /*#__PURE__*/React.createElement("span", {
    className: styles.title
  }, title), icon, count ? /*#__PURE__*/React.createElement(Badge, {
    count: count
  }) : null)), !count && newItem ? /*#__PURE__*/React.createElement(NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined
  }) : null);
});
MobileMenuItem.defaultProps = {
  underlined: true
};
export default MobileMenuItem;
