// @platform-ui/navigation/esm/MenuItemWithDropdown/MenuItemWithDropdown.js
import React, { useEffect, useMemo, useState } from 'react';
import ContextMenu from '@platform-ui/contextMenu';
import { isTouchDevice } from '@platform-ui/device-info';
import { menuItemToContextMenu } from '../utils';
import styles from './MenuItemWithDropdown.css';

function MenuItemWithDropdown({
  children,
  dropdownItems = [],
  onClick
}) {
  const [touchDevice, setTouchDevice] = useState(false);
  const memoizedItems = useMemo(() => {
    return dropdownItems == null ? void 0 : dropdownItems.map(item => menuItemToContextMenu(item, onClick));
  }, [dropdownItems, onClick]);
  useEffect(() => {
    setTouchDevice(isTouchDevice());
  }, []);
  return memoizedItems.length ? /*#__PURE__*/React.createElement(ContextMenu, {
    items: memoizedItems,
    showOnHover: true,
    align: "start",
    itemSize: "xs"
  }, children, touchDevice ? /*#__PURE__*/React.createElement("span", {
    className: styles.holder,
    tabIndex: -1,
    role: "presentation"
  }) : null) : children;
}

export default MenuItemWithDropdown;
