// @platform-ui/styleProps/borderRadius/index.js
"use strict";

exports.__esModule = true;
exports.useBorderRadiusStyle = exports.useBorderRadiusProps = exports.pickBorderRadiusProps = exports.calcBorderRadiusStyle = void 0;

var _react = require("react");

var _utils = require("../utils");

var pickBorderRadiusProps = (0, _utils.createPropsPicker)(['borderRadius', 'borderTopRadius', 'borderLeftRadius', 'borderRightRadius', 'borderBottomRadius', 'borderTopLeftRadius', 'borderTopRightRadius', 'borderBottomLeftRadius', 'borderBottomRightRadius']);
exports.pickBorderRadiusProps = pickBorderRadiusProps;

var calcBorderRadiusStyle = function calcBorderRadiusStyle(props) {
  var radius = props.borderRadius,
      topRadius = props.borderTopRadius,
      leftRadius = props.borderLeftRadius,
      rightRadius = props.borderRightRadius,
      bottomRadius = props.borderBottomRadius,
      topLeftRadius = props.borderTopLeftRadius,
      topRightRadius = props.borderTopRightRadius,
      bottomLeftRadius = props.borderBottomLeftRadius,
      bottomRightRadius = props.borderBottomRightRadius;
  var radiusArray = (0, _utils.splitShorthandProperty)(radius);
  var borderTopLeftRadius = (0, _utils.pickFirstDefined)(topLeftRadius, leftRadius, topRadius, radiusArray[0]);
  var borderTopRightRadius = (0, _utils.pickFirstDefined)(topRightRadius, rightRadius, topRadius, radiusArray[1], radiusArray[0]);
  var borderBottomLeftRadius = (0, _utils.pickFirstDefined)(bottomLeftRadius, leftRadius, bottomRadius, radiusArray[3], radiusArray[1], radiusArray[0]);
  var borderBottomRightRadius = (0, _utils.pickFirstDefined)(bottomRightRadius, rightRadius, bottomRadius, radiusArray[3], radiusArray[1], radiusArray[0]);
  return {
    borderTopLeftRadius: borderTopLeftRadius,
    borderTopRightRadius: borderTopRightRadius,
    borderBottomLeftRadius: borderBottomLeftRadius,
    borderBottomRightRadius: borderBottomRightRadius
  };
};

exports.calcBorderRadiusStyle = calcBorderRadiusStyle;

var useBorderRadiusStyle = function useBorderRadiusStyle(props) {
  var borderRadius = props.borderRadius,
      borderTopRadius = props.borderTopRadius,
      borderLeftRadius = props.borderLeftRadius,
      borderRightRadius = props.borderRightRadius,
      borderBottomRadius = props.borderBottomRadius,
      borderTopLeftRadius = props.borderTopLeftRadius,
      borderTopRightRadius = props.borderTopRightRadius,
      borderBottomLeftRadius = props.borderBottomLeftRadius,
      borderBottomRightRadius = props.borderBottomRightRadius;
  return (0, _react.useMemo)(function () {
    return calcBorderRadiusStyle({
      borderRadius: borderRadius,
      borderTopRadius: borderTopRadius,
      borderLeftRadius: borderLeftRadius,
      borderRightRadius: borderRightRadius,
      borderBottomRadius: borderBottomRadius,
      borderTopLeftRadius: borderTopLeftRadius,
      borderTopRightRadius: borderTopRightRadius,
      borderBottomLeftRadius: borderBottomLeftRadius,
      borderBottomRightRadius: borderBottomRightRadius
    });
  }, [borderRadius, borderTopRadius, borderLeftRadius, borderRightRadius, borderBottomRadius, borderTopLeftRadius, borderTopRightRadius, borderBottomLeftRadius, borderBottomRightRadius]);
};

exports.useBorderRadiusStyle = useBorderRadiusStyle;
var useBorderRadiusProps = (0, _utils.createPropsHook)(pickBorderRadiusProps, useBorderRadiusStyle);
exports.useBorderRadiusProps = useBorderRadiusProps;
