// @platform-ui/navigation/esm/Menu/MenuItem.js
import React from 'react';
import cn from 'classnames';
import Link from '../Link';
import Badge, { NewSection } from '../Badge';
import MoreButton from './MoreButton';
import MenuItemWithDropdown from '../MenuItemWithDropdown/MenuItemWithDropdown';
import styles from './MenuItem.module.css';
import { useNavigationContext } from '../NavigationContext';

const MenuItem = props => {
  const {
    title,
    active,
    underlined,
    path,
    icon,
    count,
    contextItems,
    onClick,
    dataQaType,
    dataQaData,
    hidden,
    newItem,
    dropdownItems
  } = props;
  const {
    theme
  } = useNavigationContext();
  return /*#__PURE__*/React.createElement("span", {
    className: styles.root
  }, /*#__PURE__*/React.createElement(MenuItemWithDropdown, {
    dropdownItems: dropdownItems,
    onClick: onClick
  }, /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Link, {
    className: cn(styles.link, {
      [styles.link_active]: active,
      [styles[`link_theme_${theme}`]]: theme,
      [styles.link_underlined]: underlined
    }),
    "aria-current": active ? 'page' : undefined,
    path: path,
    tabIndex: hidden ? -1 : undefined,
    onClick: onClick,
    dataQaType: dataQaType || 'uikit/navigation.menuItem',
    dataQaData: dataQaData
  }, /*#__PURE__*/React.createElement("span", {
    className: styles.title
  }, title), icon ? /*#__PURE__*/React.createElement("span", {
    className: styles.icon
  }, icon) : null, count ? /*#__PURE__*/React.createElement("span", {
    className: styles.badge
  }, /*#__PURE__*/React.createElement(Badge, {
    count: count
  })) : null), ' ')), contextItems ? /*#__PURE__*/React.createElement("span", {
    className: styles.more
  }, /*#__PURE__*/React.createElement(MoreButton, {
    items: contextItems
  })) : null, !count && newItem ? /*#__PURE__*/React.createElement(NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined
  }) : null);
};

MenuItem.defaultProps = {
  underlined: true
};
export default MenuItem;
