// @platform-ui/navigation/esm/Menu/MoreButton.js
import React, { useMemo } from 'react';
import ContextMenu from '@platform-ui/contextMenu';
import MoreHorizontal from '@platform-ui/iconsPack/interface/24/MoreHorizontal';
import dataQa from '@platform-ui/dataQa';
import cn from 'classnames';
import { menuItemToContextMenu } from '../utils';
import styles from './MoreButton.module.css';
import { useNavigationContext } from '../NavigationContext';

const MoreButton = props => {
  const {
    items
  } = props;
  const memoizedItems = useMemo(() => {
    return items.map(item => menuItemToContextMenu(item));
  }, [items]);
  const {
    theme
  } = useNavigationContext();
  return /*#__PURE__*/React.createElement("span", {
    className: cn(styles.root, {
      [styles[`root_theme_${theme}`]]: theme
    })
  }, /*#__PURE__*/React.createElement(ContextMenu, Object.assign({
    showOnHover: true,
    highlightActive: true,
    hideOnResize: true,
    items: memoizedItems,
    itemSize: "xs"
  }, dataQa(props)), /*#__PURE__*/React.createElement(MoreHorizontal, null)));
};

export default MoreButton;
