// @platform-ui/text/Text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _TextModule = _interopRequireDefault(require("./Text.module.css"));

var _excluded = ["children", "uppercase", "tagName", "size", "sizeTabletL", "sizeTabletS", "sizeDesktopS", "sizeDesktopM", "sizeDesktopL", "align", "alignTabletS", "alignTabletL", "alignDesktopS", "alignDesktopM", "alignDesktopL", "overflowEllipsis", "verticalAlign", "bold", "display", "lineThrough", "cobrowsingSecure", "whiteSpace", "wordWrap", "capitalize", "cursor", "color", "dataQaType"];

var Text = function Text(_ref) {
  var _classNames;

  var children = _ref.children,
      uppercase = _ref.uppercase,
      tagName = _ref.tagName,
      size = _ref.size,
      sizeTabletL = _ref.sizeTabletL,
      sizeTabletS = _ref.sizeTabletS,
      sizeDesktopS = _ref.sizeDesktopS,
      sizeDesktopM = _ref.sizeDesktopM,
      sizeDesktopL = _ref.sizeDesktopL,
      align = _ref.align,
      alignTabletS = _ref.alignTabletS,
      alignTabletL = _ref.alignTabletL,
      alignDesktopS = _ref.alignDesktopS,
      alignDesktopM = _ref.alignDesktopM,
      alignDesktopL = _ref.alignDesktopL,
      overflowEllipsis = _ref.overflowEllipsis,
      verticalAlign = _ref.verticalAlign,
      bold = _ref.bold,
      display = _ref.display,
      lineThrough = _ref.lineThrough,
      cobrowsingSecure = _ref.cobrowsingSecure,
      whiteSpace = _ref.whiteSpace,
      wordWrap = _ref.wordWrap,
      capitalize = _ref.capitalize,
      cursor = _ref.cursor,
      color = _ref.color,
      dataQaType = _ref.dataQaType,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(tagName || 'div', (0, _extends2.default)({}, otherProps, {
    style: {
      whiteSpace: whiteSpace,
      verticalAlign: verticalAlign,
      display: display,
      wordWrap: wordWrap,
      cursor: cursor,
      color: color
    },
    className: (0, _classnames.default)((_classNames = {}, _classNames[_TextModule.default.text] = true, _classNames[_TextModule.default["text_size_" + size]] = true, _classNames[_TextModule.default["text_sizeTabletS_" + (sizeTabletS || size)]] = true, _classNames[_TextModule.default["text_sizeTabletL_" + (sizeTabletL || sizeTabletS || size)]] = true, _classNames[_TextModule.default["text_sizeDesktopS_" + (sizeDesktopS || sizeTabletL || sizeTabletS || size)]] = true, _classNames[_TextModule.default["text_sizeDesktopM_" + (sizeDesktopM || sizeDesktopS || sizeTabletL || sizeTabletS || size)]] = true, _classNames[_TextModule.default["text_sizeDesktopL_" + (sizeDesktopL || sizeDesktopM || sizeDesktopS || sizeTabletL || sizeTabletS || size)]] = true, _classNames[_TextModule.default["text_align_" + align]] = align, _classNames[_TextModule.default["text_alignTabletS_" + alignTabletS]] = alignTabletS, _classNames[_TextModule.default["text_alignTabletL_" + alignTabletL]] = alignTabletL, _classNames[_TextModule.default["text_alignDesktopS_" + alignDesktopS]] = alignDesktopS, _classNames[_TextModule.default["text_alignDesktopM_" + alignDesktopM]] = alignDesktopM, _classNames[_TextModule.default["text_alignDesktopL_" + alignDesktopL]] = alignDesktopL, _classNames[_TextModule.default.text_bold] = bold, _classNames[_TextModule.default.text_overflowEllipsis] = overflowEllipsis, _classNames[_TextModule.default.text_lineThrough] = lineThrough, _classNames[_TextModule.default.text_uppercase] = uppercase, _classNames[_TextModule.default.text_capitalize] = capitalize, _classNames))
  }, cobrowsingSecure ? {
    'data-cobrowsing-secure': 'text'
  } : {}, (0, _dataQa.default)({
    dataQaType: dataQaType
  })), children);
};

Text.defaultProps = {
  dataQaType: 'uikit/text',
  bold: false,
  capitalize: false,
  cobrowsingSecure: false,
  lineThrough: false,
  overflowEllipsis: false,
  size: 13,
  tagName: 'div',
  uppercase: false
};
var _default = Text;
exports.default = _default;
