// @platform-ui/tabsBlock/TabsBlockCompact.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _contextMenu = _interopRequireDefault(require("@platform-ui/contextMenu"));

var _ArrowDown = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/ArrowDown"));

var _TabsBlockButton = _interopRequireDefault(require("./TabsBlockButton/TabsBlockButton"));

var _utils = require("./utils");

var _TabsBlockModule = _interopRequireDefault(require("./TabsBlock.module.css"));

var TabsBlockCompact = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2.default)(TabsBlockCompact, _PureComponent);

  function TabsBlockCompact() {
    return _PureComponent.apply(this, arguments) || this;
  }

  var _proto = TabsBlockCompact.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        size = _this$props.size,
        activeIndex = _this$props.activeIndex,
        items = _this$props.items,
        disabled = _this$props.disabled;
    var _items$activeIndex = items[activeIndex],
        title = _items$activeIndex.title,
        additionalContent = _items$activeIndex.additionalContent;
    return /*#__PURE__*/_react.default.createElement(_contextMenu.default, {
      wide: true,
      colored: true,
      hideOnResize: true,
      items: (0, _utils.getContextMenuItems)(this.props).filter(function (item) {
        return !item.active;
      })
    }, /*#__PURE__*/_react.default.createElement(_TabsBlockButton.default, {
      wide: true,
      first: true,
      last: true,
      disabled: disabled,
      size: size
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _TabsBlockModule.default.moreButtonContent
    }, title), additionalContent && /*#__PURE__*/_react.default.createElement("div", {
      className: _TabsBlockModule.default.additionalContent
    }, additionalContent), /*#__PURE__*/_react.default.createElement("div", {
      className: _TabsBlockModule.default.icon
    }, /*#__PURE__*/_react.default.createElement(_ArrowDown.default, null))));
  };

  return TabsBlockCompact;
}(_react.PureComponent);

TabsBlockCompact.defaultProps = {
  size: 'm',
  activeIndex: 0,
  items: []
};
var _default = TabsBlockCompact;
exports.default = _default;
