// @platform-ui/tabsBlock/utils.js
"use strict";

exports.__esModule = true;
exports.getOnClickHandler = exports.getKey = exports.getContextMenuItems = void 0;

var getOnClickHandler = function getOnClickHandler(handlers, activeIndex, title) {
  return function (e) {
    return handlers.forEach(function (handler) {
      return handler && handler(e, {
        activeIndex: activeIndex,
        title: title
      });
    });
  };
};

exports.getOnClickHandler = getOnClickHandler;

var getContextMenuItems = function getContextMenuItems(_ref) {
  var activeIndex = _ref.activeIndex,
      onItemClick = _ref.onItemClick,
      items = _ref.items,
      disabled = _ref.disabled;
  return items.map(function (item, index) {
    return {
      additionalContent: item.additionalContent,
      disabled: disabled || item.disabled,
      title: item.title,
      onClick: getOnClickHandler([onItemClick, item.onClick], index, item.title),
      active: index === activeIndex
    };
  });
};

exports.getContextMenuItems = getContextMenuItems;

var getKey = function getKey(item, index) {
  return item.key || item.title || (typeof item.additionalContent === 'string' ? item.additionalContent : undefined) || index;
};

exports.getKey = getKey;
