// @platform-ui/tabsBlock/esm/TabsBlockCompact.js
import React, { PureComponent } from 'react';
import ContextMenu from '@platform-ui/contextMenu';
import ArrowIcon from '@platform-ui/iconsPack/interface/24/ArrowDown';
import TabsBlockButton from './TabsBlockButton/TabsBlockButton';
import { getContextMenuItems } from './utils';
import styles from './TabsBlock.module.css';

class TabsBlockCompact extends PureComponent {
  render() {
    const {
      size,
      activeIndex,
      items,
      disabled
    } = this.props;
    const {
      title,
      additionalContent
    } = items[activeIndex];
    return /*#__PURE__*/React.createElement(ContextMenu, {
      wide: true,
      colored: true,
      hideOnResize: true,
      items: getContextMenuItems(this.props).filter(item => !item.active)
    }, /*#__PURE__*/React.createElement(TabsBlockButton, {
      wide: true,
      first: true,
      last: true,
      disabled: disabled,
      size: size
    }, /*#__PURE__*/React.createElement("div", {
      className: styles.moreButtonContent
    }, title), additionalContent && /*#__PURE__*/React.createElement("div", {
      className: styles.additionalContent
    }, additionalContent), /*#__PURE__*/React.createElement("div", {
      className: styles.icon
    }, /*#__PURE__*/React.createElement(ArrowIcon, null))));
  }

}

TabsBlockCompact.defaultProps = {
  size: 'm',
  activeIndex: 0,
  items: []
};
export default TabsBlockCompact;
