// @platform-ui/styleProps/esm/margin/index.js
import { useMemo } from 'react';
import { pickFirstDefined, splitShorthandProperty, createPropsPicker, createPropsHook } from '../utils';
export const pickMarginProps = createPropsPicker(['margin', 'marginX', 'marginY', 'marginTop', 'marginLeft', 'marginRight', 'marginBottom']);
export const calcMarginStyle = props => {
  const {
    margin: m,
    marginX: mX,
    marginY: mY,
    marginTop: mTop,
    marginLeft: mLeft,
    marginRight: mRight,
    marginBottom: mBottom
  } = props;
  const mA = splitShorthandProperty(m);
  const marginTop = pickFirstDefined(mTop, mY, mA[0]);
  const marginLeft = pickFirstDefined(mLeft, mX, mA[3], mA[1], mA[0]);
  const marginRight = pickFirstDefined(mRight, mX, mA[1], mA[0]);
  const marginBottom = pickFirstDefined(mBottom, mY, mA[2], mA[0]);
  return {
    marginTop,
    marginLeft,
    marginRight,
    marginBottom
  };
};
export const useMarginStyle = props => {
  const {
    margin,
    marginX,
    marginY,
    marginTop,
    marginLeft,
    marginRight,
    marginBottom
  } = props;
  return useMemo(() => calcMarginStyle({
    margin,
    marginX,
    marginY,
    marginTop,
    marginLeft,
    marginRight,
    marginBottom
  }), [margin, marginX, marginY, marginTop, marginLeft, marginRight, marginBottom]);
};
export const useMarginProps = createPropsHook(pickMarginProps, useMarginStyle);
