// @platform-ui/popup/Paper/Paper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _styleProps = require("@platform-ui/styleProps");

var _PaperModule = _interopRequireDefault(require("./Paper.module.css"));

var Paper = function Paper(props) {
  var _cn;

  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/popup.paper' : _props$dataQaType,
      _props$size = props.size,
      size = _props$size === void 0 ? 'l' : _props$size,
      _props$mobileSize = props.mobileSize,
      mobileSize = _props$mobileSize === void 0 ? 's' : _props$mobileSize,
      media = props.media,
      disableHorizontalPadding = props.disableHorizontalPadding,
      children = props.children,
      backgroundColor = props.backgroundColor,
      _props$alignCross = props.alignCross,
      alignCross = _props$alignCross === void 0 ? 'center' : _props$alignCross,
      minWidth = props.minWidth,
      gap = props.gap,
      mobileGap = props.mobileGap;
  var paddingStyle = (0, _styleProps.usePaddingStyle)(props);
  var customSize = typeof size === 'number';
  return /*#__PURE__*/_react.default.createElement("div", Object.assign({
    className: (0, _classnames.default)(_PaperModule.default.root, (_cn = {}, _cn[_PaperModule.default["root_size_" + (customSize ? 'custom' : size)]] = size, _cn[_PaperModule.default["root_media_" + media]] = media, _cn[_PaperModule.default["root_mobileSize_" + mobileSize]] = mobileSize, _cn[_PaperModule.default.root_disableHorizontalPadding] = disableHorizontalPadding, _cn[_PaperModule.default["root_alignCross_" + alignCross]] = alignCross, _cn)),
    style: {
      padding: media === 'phone' ? mobileGap : gap
    }
  }, (0, _dataQa.default)({
    dataQaType: dataQaType
  })), /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".container"
  }), {
    className: _PaperModule.default.paper,
    style: (0, _extends2.default)({}, paddingStyle, {
      maxWidth: customSize ? size + "px" : undefined,
      backgroundColor: backgroundColor,
      minWidth: minWidth
    })
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _PaperModule.default.paperInner
  }, children)));
};

var _default = Paper;
exports.default = _default;
