// @platform-ui/block/esm/withBlock.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["dataQaType", "disabled", "icon", "title", "label", "description", "hintBlock", "size", "vertical", "reverse", "inGroup", "separated", "flatCorners", "borderType", "alignSections", "children"];
import React from 'react';
import Block from './Block';
const INPUT_SIZE_MAP = {
  s: 's',
  m: 's',
  l: 'l',
  xl: 'l'
};
export const withBlock = (Component, defaults = {}) => props => {
  const {
    dataQaType = defaults.dataQaType,
    disabled,
    icon,
    title,
    label,
    description,
    hintBlock,
    size = 'l',
    vertical,
    reverse,
    inGroup,
    separated,
    flatCorners,
    borderType,
    alignSections,
    children
  } = props,
        inputProps = _objectWithoutPropertiesLoose(props, _excluded);

  const input = /*#__PURE__*/React.createElement(Component, Object.assign({}, inputProps, {
    disabled: disabled,
    size: INPUT_SIZE_MAP[size],
    dataQaType: `${dataQaType}.input`
  }));
  return /*#__PURE__*/React.createElement(Block, {
    dataQaType: dataQaType,
    size: size,
    disabled: disabled,
    input: input,
    icon: icon,
    title: title,
    label: label,
    description: description,
    hintBlock: hintBlock,
    vertical: vertical,
    reverse: reverse,
    inGroup: inGroup,
    separated: separated,
    flatCorners: flatCorners,
    borderType: borderType,
    alignSections: alignSections
  }, children);
};
