// @platform-ui/navigation/User/Logout/Logout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Logout = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/Logout"));

var _Link = _interopRequireDefault(require("../../Link"));

var _LogoutModule = _interopRequireDefault(require("./Logout.module.css"));

var Logout = function Logout(props) {
  var _cn;

  var title = props.title,
      path = props.path,
      onClick = props.onClick,
      mobile = props.mobile,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData;
  return /*#__PURE__*/_react.default.createElement(_Link.default, {
    className: (0, _classnames.default)(_LogoutModule.default.root, (_cn = {}, _cn[_LogoutModule.default.root_mobile] = mobile, _cn)),
    path: path,
    onClick: onClick,
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _LogoutModule.default.title
  }, title), /*#__PURE__*/_react.default.createElement("span", {
    className: _LogoutModule.default.icon
  }, /*#__PURE__*/_react.default.createElement(_Logout.default, null)));
};

var _default = Logout;
exports.default = _default;
