// @platform-ui/navigation/esm/User/UserDesktop/UserDesktop.js
import React from 'react';
import dataQa from '@platform-ui/dataQa';
import Logout from '../Logout';
import Profile from '../Profile';
import CompaniesList from '../Companies';
import { Divider } from '../../index';
import UserDesktopRoot from './UserDesktopRoot';

const UserDesktop = props => {
  var _user$companies;

  const {
    user,
    logoutLink,
    onClick,
    menuComponent
  } = props;
  return /*#__PURE__*/React.createElement(UserDesktopRoot, Object.assign({
    menuComponent: menuComponent
  }, dataQa(props)), /*#__PURE__*/React.createElement(CompaniesList, {
    companies: user.companies,
    onClick: onClick,
    profileLink: user.profileLink
  }), (_user$companies = user.companies) != null && _user$companies.length ? /*#__PURE__*/React.createElement(Divider, null) : null, /*#__PURE__*/React.createElement(Profile, {
    avatarUrl: user.avatarUrl,
    fullName: user.fullName,
    premium: user.premium,
    links: user.profileSections,
    onClick: onClick,
    badgeSubscriptions: user.badgeSubscriptions
  }), /*#__PURE__*/React.createElement(Logout, Object.assign({
    onClick: onClick
  }, logoutLink)));
};

export default UserDesktop;
