// @platform-ui/popup/Controls/CloseButton/CloseButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _Close = _interopRequireDefault(require("@platform-ui/iconsPack/interface/16/Close"));

var _CloseButtonModule = _interopRequireDefault(require("./CloseButton.module.css"));

var CloseButton = function CloseButton(props) {
  var _cn;

  var onClick = props.onClick,
      theme = props.theme,
      media = props.media,
      dataQaType = props.dataQaType,
      _props$ariaLabel = props.ariaLabel,
      ariaLabel = _props$ariaLabel === void 0 ? 'Закрыть попап' : _props$ariaLabel;

  var handleMouseDown = function handleMouseDown(event) {
    event.preventDefault();
  };

  return /*#__PURE__*/_react.default.createElement("button", Object.assign({
    type: "button",
    className: (0, _classnames.default)(_CloseButtonModule.default.root, (_cn = {}, _cn[_CloseButtonModule.default["root_theme_" + theme]] = theme, _cn[_CloseButtonModule.default["root_media_" + media]] = media, _cn)),
    onClick: onClick,
    onMouseDown: handleMouseDown,
    "aria-label": ariaLabel
  }, (0, _dataQa.default)({
    dataQaType: dataQaType + ".closeButton"
  })), /*#__PURE__*/_react.default.createElement(_Close.default, null));
};

var _default = CloseButton;
exports.default = _default;
