// @platform-ui/money/Money.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = exports.PropTypes = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _utils = require("./utils");

var _currencies = _interopRequireDefault(require("./currencies.json"));

var _MoneyModule = _interopRequireDefault(require("./Money.module.css"));

var _excluded = ["color", "cobrowsingSecure", "digitGroupingSeparator", "decimalSeparator", "symbolPosition", "children", "dataQaData", "dataQaType", "value", "showPlus", "styledSmalls", "showSmalls", "showCurrencyWithNull", "trimZeros"];
var PropTypes = {
  currency: Object.keys(_currencies.default)
};
exports.PropTypes = PropTypes;

var Money = function Money(_ref) {
  var _cn;

  var color = _ref.color,
      cobrowsingSecure = _ref.cobrowsingSecure,
      _ref$digitGroupingSep = _ref.digitGroupingSeparator,
      digitGroupingSeparator = _ref$digitGroupingSep === void 0 ? " " : _ref$digitGroupingSep,
      _ref$decimalSeparator = _ref.decimalSeparator,
      decimalSeparator = _ref$decimalSeparator === void 0 ? ',' : _ref$decimalSeparator,
      _ref$symbolPosition = _ref.symbolPosition,
      symbolPosition = _ref$symbolPosition === void 0 ? 'postfix' : _ref$symbolPosition,
      children = _ref.children,
      dataQaData = _ref.dataQaData,
      _ref$dataQaType = _ref.dataQaType,
      dataQaType = _ref$dataQaType === void 0 ? 'uikit/money' : _ref$dataQaType,
      _ref$value = _ref.value,
      value = _ref$value === void 0 ? 0 : _ref$value,
      _ref$showPlus = _ref.showPlus,
      showPlus = _ref$showPlus === void 0 ? false : _ref$showPlus,
      _ref$styledSmalls = _ref.styledSmalls,
      styledSmalls = _ref$styledSmalls === void 0 ? true : _ref$styledSmalls,
      _ref$showSmalls = _ref.showSmalls,
      showSmalls = _ref$showSmalls === void 0 ? true : _ref$showSmalls,
      _ref$showCurrencyWith = _ref.showCurrencyWithNull,
      showCurrencyWithNull = _ref$showCurrencyWith === void 0 ? true : _ref$showCurrencyWith,
      _ref$trimZeros = _ref.trimZeros,
      trimZeros = _ref$trimZeros === void 0 ? true : _ref$trimZeros,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);

  var _formatMoney = (0, _utils.formatMoney)((0, _extends2.default)({
    decimalSeparator: decimalSeparator,
    showPlus: showPlus,
    showSmalls: showSmalls,
    showCurrencyWithNull: showCurrencyWithNull,
    trimZeros: trimZeros,
    value: value
  }, otherProps)),
      sign = _formatMoney.sign,
      numerals = _formatMoney.numerals,
      smalls = _formatMoney.smalls,
      currency = _formatMoney.currency;

  var layout = {
    prefix: '',
    postfix: ''
  };
  layout[symbolPosition] = currency;
  return /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaData: dataQaData,
    dataQaType: dataQaType
  }), {
    "data-cobrowsing-secure": cobrowsingSecure ? 'text' : undefined,
    className: (0, _classnames.default)((_cn = {}, _cn[_MoneyModule.default.money] = true, _cn[_MoneyModule.default["money_color_" + color]] = color, _cn))
  }), sign, currency && layout.prefix, children || numerals.map(function (numeral, index) {
    return (
      /*#__PURE__*/
      // eslint-disable-next-line react/no-array-index-key
      _react.default.createElement(_react.default.Fragment, {
        key: index
      }, index !== 0 && digitGroupingSeparator, numeral)
    );
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: styledSmalls ? _MoneyModule.default.smalls : undefined
  }, smalls, currency && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "\xA0", layout.postfix)));
};

var _default = Money;
exports.default = _default;
