// @platform-ui/styleProps/esm/borderRadius/index.js
import { useMemo } from 'react';
import { pickFirstDefined, splitShorthandProperty, createPropsPicker, createPropsHook } from '../utils';
export const pickBorderRadiusProps = createPropsPicker(['borderRadius', 'borderTopRadius', 'borderLeftRadius', 'borderRightRadius', 'borderBottomRadius', 'borderTopLeftRadius', 'borderTopRightRadius', 'borderBottomLeftRadius', 'borderBottomRightRadius']);
export const calcBorderRadiusStyle = props => {
  const {
    borderRadius: radius,
    borderTopRadius: topRadius,
    borderLeftRadius: leftRadius,
    borderRightRadius: rightRadius,
    borderBottomRadius: bottomRadius,
    borderTopLeftRadius: topLeftRadius,
    borderTopRightRadius: topRightRadius,
    borderBottomLeftRadius: bottomLeftRadius,
    borderBottomRightRadius: bottomRightRadius
  } = props;
  const radiusArray = splitShorthandProperty(radius);
  const borderTopLeftRadius = pickFirstDefined(topLeftRadius, leftRadius, topRadius, radiusArray[0]);
  const borderTopRightRadius = pickFirstDefined(topRightRadius, rightRadius, topRadius, radiusArray[1], radiusArray[0]);
  const borderBottomLeftRadius = pickFirstDefined(bottomLeftRadius, leftRadius, bottomRadius, radiusArray[3], radiusArray[1], radiusArray[0]);
  const borderBottomRightRadius = pickFirstDefined(bottomRightRadius, rightRadius, bottomRadius, radiusArray[3], radiusArray[1], radiusArray[0]);
  return {
    borderTopLeftRadius,
    borderTopRightRadius,
    borderBottomLeftRadius,
    borderBottomRightRadius
  };
};
export const useBorderRadiusStyle = props => {
  const {
    borderRadius,
    borderTopRadius,
    borderLeftRadius,
    borderRightRadius,
    borderBottomRadius,
    borderTopLeftRadius,
    borderTopRightRadius,
    borderBottomLeftRadius,
    borderBottomRightRadius
  } = props;
  return useMemo(() => calcBorderRadiusStyle({
    borderRadius,
    borderTopRadius,
    borderLeftRadius,
    borderRightRadius,
    borderBottomRadius,
    borderTopLeftRadius,
    borderTopRightRadius,
    borderBottomLeftRadius,
    borderBottomRightRadius
  }), [borderRadius, borderTopRadius, borderLeftRadius, borderRightRadius, borderBottomRadius, borderTopLeftRadius, borderTopRightRadius, borderBottomLeftRadius, borderBottomRightRadius]);
};
export const useBorderRadiusProps = createPropsHook(pickBorderRadiusProps, useBorderRadiusStyle);
