// @platform-ui/input/InputHintHOC.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _generateId = require("@platform-ui/generateId");

var _tooltip = require("@platform-ui/tooltip");

var _excluded = ["hintContent", "id", "isActiveHint", "isInactiveHint", "focusableHint", "hintBubbleTheme"];

var InputHintHOC = function InputHintHOC(WrappedComponent) {
  var getName = function getName(Component) {
    return Component.displayName || Component.name || 'Component(!)';
  };

  var HintWrapper = function HintWrapper(_ref) {
    var hintContent = _ref.hintContent,
        propsId = _ref.id,
        isActiveHint = _ref.isActiveHint,
        isInactiveHint = _ref.isInactiveHint,
        focusableHint = _ref.focusableHint,
        hintBubbleTheme = _ref.hintBubbleTheme,
        otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
    var id = (0, _generateId.useId)(propsId);

    var handleMouseDown = function handleMouseDown(event) {
      event.stopPropagation();
    };

    return /*#__PURE__*/_react.default.createElement(WrappedComponent, Object.assign({}, otherProps, {
      id: id,
      hintBlock: hintContent ? /*#__PURE__*/_react.default.createElement("span", {
        onMouseDown: handleMouseDown
      }, /*#__PURE__*/_react.default.createElement(_tooltip.Hint, {
        popoverContent: hintContent,
        direction: "bottom",
        align: "end",
        theme: hintBubbleTheme,
        additionalShift: 12,
        ariaOwnsId: id,
        isActive: isActiveHint,
        isInactive: isInactiveHint,
        focusable: focusableHint
      })) : undefined
    }));
  };

  process.env.NODE_ENV !== "production" ? HintWrapper.propTypes = (0, _extends2.default)({}, WrappedComponent.propTypes, {
    hintContent: _propTypes.default.node,
    isActiveHint: _propTypes.default.bool,
    isInactiveHint: _propTypes.default.bool,
    hintBubbleTheme: _propTypes.default.string
  }) : void 0;
  HintWrapper.defaultProps = (0, _extends2.default)({}, WrappedComponent.defaultProps, {
    hintContent: null,
    isActiveHint: undefined,
    isInactiveHint: undefined,
    hintBubbleTheme: 'dark'
  });
  HintWrapper.displayName = getName(WrappedComponent);
  return (0, _hoistNonReactStatics.default)(HintWrapper, WrappedComponent);
};

var _default = InputHintHOC;
exports.default = _default;
