// @platform-ui/inputTools/autofillHOC.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _autofillHOCModule = _interopRequireDefault(require("./autofillHOC.module.css"));

var autofillHOC = function autofillHOC(WrappedComponent) {
  var WithAutofill = function WithAutofill(props) {
    var _useState = (0, _react.useState)(false),
        autofilled = _useState[0],
        setAutofilled = _useState[1];

    var handleAnimationStart = function handleAnimationStart(event) {
      if (event.animationName === _autofillHOCModule.default.onAutoFillStart) {
        setAutofilled(true);
      }

      if (event.animationName === _autofillHOCModule.default.onAutoFillCancel) {
        setAutofilled(false);
      }
    };

    return /*#__PURE__*/_react.default.createElement(WrappedComponent, Object.assign({}, props, autofilled ? {
      filled: true,
      placeholder: ''
    } : {}, {
      autoComplete: true,
      autofilled: autofilled,
      onAnimationStart: handleAnimationStart
    }));
  };

  return (0, _hoistNonReactStatics.default)(WithAutofill, WrappedComponent);
};

var _default = autofillHOC;
exports.default = _default;
