// @platform-ui/inputTools/InlineInput.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _InlineInputModule = _interopRequireDefault(require("./InlineInput.module.css"));

var _autofillHOCModule = _interopRequireDefault(require("./autofillHOC.module.css"));

var _excluded = ["minWidth", "autoComplete", "value", "placeholder", "cobrowsingSecure", "dataQaType", "refCallback", "ellipsis"];
var ZERO_WIDTH_SPACE = "\u200B";

var autoCompleteValue = function autoCompleteValue(autoCompleteProp) {
  if (typeof autoCompleteProp === 'string') {
    return autoCompleteProp;
  }

  return autoCompleteProp ? 'on' : 'off';
};

var InlineInput = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _cn;

  var minWidth = _ref.minWidth,
      autoComplete = _ref.autoComplete,
      value = _ref.value,
      placeholder = _ref.placeholder,
      cobrowsingSecure = _ref.cobrowsingSecure,
      dataQaType = _ref.dataQaType,
      refCallback = _ref.refCallback,
      ellipsis = _ref.ellipsis,
      props = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var valuePlaceholder = value || placeholder;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: _InlineInputModule.default.container,
    style: {
      minWidth: minWidth
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _InlineInputModule.default.value,
    "data-qa-type": dataQaType + ".text",
    "data-cobrowsing-secure": cobrowsingSecure ? 'text' : undefined
  }, valuePlaceholder || ZERO_WIDTH_SPACE), /*#__PURE__*/_react.default.createElement("input", Object.assign({}, props, {
    "data-qa-type": dataQaType + ".input",
    "data-cobrowsing-secure": cobrowsingSecure ? 'input' : undefined,
    autoComplete: autoComplete !== undefined ? autoCompleteValue(autoComplete) : undefined,
    className: (0, _classnames.default)((_cn = {}, _cn[_InlineInputModule.default.input] = true, _cn[_autofillHOCModule.default.input] = true, _cn[_InlineInputModule.default.input_ellipsis] = ellipsis, _cn)),
    ref: ref || refCallback,
    value: value
  })));
});
InlineInput.displayName = 'InlineInput';
InlineInput.defaultProps = {
  dataQaType: 'uikit/inlineInput',
  cobrowsingSecure: false
};
var _default = InlineInput;
exports.default = _default;
