// @platform-ui/inputTools/notNullableNumberInputHOC.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _memoizeWeak = _interopRequireDefault(require("@platform-ui/memoizeWeak"));

var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));

var _parse = require("./numberInputUtils/parse");

var notNullableNumberInputHOC = (0, _memoizeWeak.default)(function (NumberInput) {
  var NotNullableInput = /*#__PURE__*/function (_Component) {
    (0, _inheritsLoose2.default)(NotNullableInput, _Component);

    function NotNullableInput(props) {
      var _this;

      _this = _Component.call(this, props) || this;

      _this.handleChange = function (event, params) {
        var value = params.value;

        if ((0, _parse.isEmptyValue)(value)) {
          value = _this.state.emptyValue;

          _this.setState({
            empty: true
          });
        }

        if (value === _this.props.value) {
          return;
        }

        _this.props.onChange(event, (0, _extends2.default)({}, params, {
          value: value
        }));
      };

      _this.state = {
        empty: false,
        emptyValue: (0, _parse.getEmptyValue)(props)
      };
      return _this;
    }

    NotNullableInput.getDerivedStateFromProps = function getDerivedStateFromProps(props, state) {
      var emptyValue = (0, _parse.getEmptyValue)(props);
      var empty = state.empty && props.focused && props.value === emptyValue;
      return {
        empty: empty,
        emptyValue: emptyValue
      };
    };

    var _proto = NotNullableInput.prototype;

    _proto.render = function render() {
      var _this$props = this.props,
          value = _this$props.value,
          forwardedRef = _this$props.forwardedRef;
      var empty = this.state.empty;
      return /*#__PURE__*/_react.default.createElement(NumberInput, Object.assign({}, this.props, {
        ref: forwardedRef,
        value: empty ? _parse.EMPTY_VALUE : value,
        onChange: this.handleChange
      }));
    };

    return NotNullableInput;
  }(_react.Component);

  (0, _hoistNonReactStatics.default)(NotNullableInput, NumberInput);
  return /*#__PURE__*/(0, _react.forwardRef)(function (props, forwardedRef) {
    return /*#__PURE__*/_react.default.createElement(NotNullableInput, Object.assign({}, props, {
      forwardedRef: forwardedRef
    }));
  });
});
var _default = notNullableNumberInputHOC;
exports.default = _default;
