// @platform-ui/inputTools/placeholderHOC.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _excluded = ["value", "placeholder", "partialPlaceholder", "prefix", "__overrideStyles", "whiteSpace"];

var placeholderHOC = function placeholderHOC(Placeholder) {
  return function (Input) {
    return /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
      var value = _ref.value,
          placeholder = _ref.placeholder,
          partialPlaceholder = _ref.partialPlaceholder,
          prefix = _ref.prefix,
          __overrideStyles = _ref.__overrideStyles,
          whiteSpace = _ref.whiteSpace,
          otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
      return /*#__PURE__*/_react.default.createElement(Placeholder, {
        dataQaType: otherProps.dataQaType && otherProps.dataQaType + ".placeholder",
        value: value,
        placeholder: placeholder,
        partialPlaceholder: partialPlaceholder,
        prefix: prefix,
        minWidth: otherProps.minWidth,
        __overrideStyles: __overrideStyles,
        whiteSpace: whiteSpace,
        input: /*#__PURE__*/_react.default.createElement(Input, Object.assign({
          ref: ref,
          value: value,
          prefix: prefix,
          placeholder: placeholder
        }, otherProps))
      });
    });
  };
};

var _default = placeholderHOC;
exports.default = _default;
