// @platform-ui/navigation/User/UserMobile/UserMobile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Logout = _interopRequireDefault(require("../Logout"));

var _Profile = _interopRequireDefault(require("../Profile"));

var _Companies = _interopRequireDefault(require("../Companies"));

var _index = require("../../index");

var _UserMobileRoot = _interopRequireDefault(require("./UserMobileRoot"));

var _UserMobileSpacer = _interopRequireDefault(require("./UserMobileSpacer"));

var UserMobile = function UserMobile(props) {
  var _user$companies;

  var user = props.user,
      logoutLink = props.logoutLink,
      onClick = props.onClick;
  return /*#__PURE__*/_react.default.createElement(_UserMobileRoot.default, null, /*#__PURE__*/_react.default.createElement(_Profile.default, {
    mobile: true,
    avatarUrl: user.avatarUrl,
    fullName: user.fullName,
    premium: user.premium,
    links: user.profileSections,
    onClick: onClick,
    badgeSubscriptions: user.badgeSubscriptions
  }), (_user$companies = user.companies) != null && _user$companies.length ? /*#__PURE__*/_react.default.createElement(_index.Divider, null) : null, /*#__PURE__*/_react.default.createElement(_Companies.default, {
    profileLink: user.profileLink,
    onClick: onClick,
    companies: user.companies,
    mobile: true
  }), /*#__PURE__*/_react.default.createElement(_UserMobileSpacer.default, null), /*#__PURE__*/_react.default.createElement(_Logout.default, Object.assign({
    onClick: onClick
  }, logoutLink, {
    mobile: true
  })));
};

var _default = UserMobile;
exports.default = _default;
