// @platform-ui/navigation/User/UserMobile/UserMobileRoot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _Menu = _interopRequireDefault(require("../Icons/Menu"));

var _MobilePopup = _interopRequireDefault(require("../../MobilePopup"));

var _UserMobileRootModule = _interopRequireDefault(require("./UserMobileRoot.module.css"));

var UserMobileRoot = function UserMobileRoot(props) {
  var profileComponent = props.children;

  var _useState = (0, _react.useState)(false),
      menuVisible = _useState[0],
      setMenuVisible = _useState[1];

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _UserMobileRootModule.default.root
  }, /*#__PURE__*/_react.default.createElement("div", {
    role: "button",
    className: _UserMobileRootModule.default.icon,
    onClick: function onClick() {
      setMenuVisible(true);
    }
  }, /*#__PURE__*/_react.default.createElement(_Menu.default, null)), /*#__PURE__*/_react.default.createElement(_MobilePopup.default, {
    active: menuVisible,
    align: "right",
    onClose: function onClose() {
      setMenuVisible(false);
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _UserMobileRootModule.default.content
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, typeof profileComponent === 'function' ? profileComponent({
    closeProfilePopover: function closeProfilePopover() {
      return setMenuVisible(false);
    }
  }) : profileComponent))));
};

var _default = UserMobileRoot;
exports.default = _default;
