// @platform-ui/navigation/AccordionMenu/AccordionMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _AccordionMenuItem = _interopRequireDefault(require("./AccordionMenuItem"));

var _AccordionMenuModule = _interopRequireDefault(require("./AccordionMenu.module.css"));

var AccordionMenu = function AccordionMenu(props) {
  var items = props.items,
      onClick = props.onClick,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      id = props.id;
  var activeIndex = (0, _react.useMemo)(function () {
    return items.findIndex(function (item) {
      return Boolean(item.active);
    });
  }, [items]);

  var _useState = (0, _react.useState)(),
      expanded = _useState[0],
      setExpanded = _useState[1];

  var handleExpand = function handleExpand(index) {
    return function (event) {
      event.stopPropagation();
      setExpanded(index === expanded ? undefined : index);
    };
  };

  (0, _react.useEffect)(function () {
    setExpanded(activeIndex);
  }, [activeIndex]);
  return /*#__PURE__*/_react.default.createElement("div", {
    id: id,
    className: _AccordionMenuModule.default.root,
    "data-qa-type": dataQaType || 'uikit/navigation.accordionMenu',
    "data-qa-data": dataQaData
  }, items.map(function (itemProps, index) {
    return /*#__PURE__*/_react.default.createElement(_AccordionMenuItem.default // eslint-disable-next-line react/no-array-index-key
    , Object.assign({
      key: index,
      onClick: onClick,
      expanded: index === expanded,
      onExpand: handleExpand(index),
      lastChild: index === items.length - 1
    }, itemProps));
  }));
};

var _default = AccordionMenu;
exports.default = _default;
