// @platform-ui/navigation/AccordionMenu/AccordionMenuItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _collapseBlock = _interopRequireDefault(require("@platform-ui/collapseBlock"));

var _ArrowDown = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/ArrowDown"));

var _Divider = _interopRequireDefault(require("../Divider"));

var _AccordionMenuLink = _interopRequireDefault(require("./AccordionMenuLink"));

var _AccordionMenuItemModule = _interopRequireDefault(require("./AccordionMenuItem.module.css"));

var _Badge = require("../Badge");

var AccordionMenuItem = function AccordionMenuItem(props) {
  var _cn, _cn2;

  var title = props.title,
      menuItems = props.items,
      expanded = props.expanded,
      alwaysExpanded = props.alwaysExpanded,
      onExpand = props.onExpand,
      onClick = props.onClick,
      lastChild = props.lastChild,
      active = props.active,
      dataQaData = props.dataQaData,
      _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/navigation.accordionMenuItem' : _props$dataQaType,
      dropdownItems = props.dropdownItems,
      newItem = props.newItem;
  var items = dropdownItems || menuItems;

  if (!(items != null && items.length)) {
    return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _dataQa.default)({
      dataQaType: dataQaType,
      dataQaData: dataQaData
    }), {
      className: _AccordionMenuItemModule.default.root
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: _AccordionMenuItemModule.default.row
    }, /*#__PURE__*/_react.default.createElement(_AccordionMenuLink.default, Object.assign({}, props, {
      active: active,
      onClick: onClick
    }))), lastChild ? null : /*#__PURE__*/_react.default.createElement(_Divider.default, null));
  }

  return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _dataQa.default)({
    dataQaType: (0, _classnames.default)(dataQaType, (_cn = {}, _cn[dataQaType + "_expanded"] = expanded, _cn)),
    dataQaData: dataQaData
  }), {
    className: (0, _classnames.default)(_AccordionMenuItemModule.default.root, (_cn2 = {}, _cn2[_AccordionMenuItemModule.default.root_expanded] = expanded, _cn2[_AccordionMenuItemModule.default.root_active] = active, _cn2))
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: _AccordionMenuItemModule.default.row,
    onClick: onExpand
  }, /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".title"
  }), {
    className: _AccordionMenuItemModule.default.title
  }), title), newItem ? /*#__PURE__*/_react.default.createElement("span", {
    className: _AccordionMenuItemModule.default.newItem
  }, /*#__PURE__*/_react.default.createElement(_Badge.NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined,
    sup: false
  })) : null, !alwaysExpanded ? /*#__PURE__*/_react.default.createElement("span", {
    className: _AccordionMenuItemModule.default.iconWrapper
  }, /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".icon"
  }), {
    className: _AccordionMenuItemModule.default.icon
  }), /*#__PURE__*/_react.default.createElement(_ArrowDown.default, null))) : null), /*#__PURE__*/_react.default.createElement(_collapseBlock.default, {
    collapsed: !alwaysExpanded && !expanded
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _AccordionMenuItemModule.default.items
  }, items.map(function (itemProps, index) {
    return /*#__PURE__*/_react.default.createElement(_AccordionMenuLink.default // eslint-disable-next-line react/no-array-index-key
    , Object.assign({
      key: index,
      onClick: onClick,
      dataQaType: itemProps.dataQaType || dataQaType + "Second"
    }, itemProps));
  }))), lastChild ? null : /*#__PURE__*/_react.default.createElement(_Divider.default, null));
};

var _default = AccordionMenuItem;
exports.default = _default;
