// @platform-ui/navigation/AccordionMenu/AccordionMenuLink.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Link = _interopRequireDefault(require("../Link"));

var _Badge = _interopRequireWildcard(require("../Badge"));

var _AccordionMenuLinkModule = _interopRequireDefault(require("./AccordionMenuLink.module.css"));

var AccordionMenuLink = function AccordionMenuLink(props) {
  var _cn;

  var title = props.title,
      active = props.active,
      path = props.path,
      icon = props.icon,
      count = props.count,
      onClick = props.onClick,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      newItem = props.newItem;
  return /*#__PURE__*/_react.default.createElement("span", {
    className: _AccordionMenuLinkModule.default.root,
    "data-qa-type": "uikit/navigation.accordionMenuLink"
  }, /*#__PURE__*/_react.default.createElement(_Link.default, {
    className: (0, _classnames.default)(_AccordionMenuLinkModule.default.link, (_cn = {}, _cn[_AccordionMenuLinkModule.default.link_active] = active, _cn)),
    "aria-current": active ? 'page' : undefined,
    path: path,
    onClick: onClick,
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _AccordionMenuLinkModule.default.title
  }, title), icon, count ? /*#__PURE__*/_react.default.createElement(_Badge.default, {
    count: count
  }) : null, !count && newItem ? /*#__PURE__*/_react.default.createElement("span", {
    className: _AccordionMenuLinkModule.default.newItem
  }, /*#__PURE__*/_react.default.createElement(_Badge.NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined,
    sup: false
  })) : null));
};

var _default = AccordionMenuLink;
exports.default = _default;
