// @platform-ui/navigation/MobileMenu/MobileMenuItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _MobileMenuItemModule = _interopRequireDefault(require("./MobileMenuItem.module.css"));

var _MenuItemWithDropdown = _interopRequireDefault(require("../MenuItemWithDropdown"));

var _Link = _interopRequireDefault(require("../Link"));

var _Badge = _interopRequireWildcard(require("../Badge"));

var _NavigationContext = require("../NavigationContext");

var MobileMenuItem = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _cn, _cn2;

  var title = props.title,
      active = props.active,
      underlined = props.underlined,
      path = props.path,
      icon = props.icon,
      count = props.count,
      onClick = props.onClick,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      newItem = props.newItem,
      dropdownItems = props.dropdownItems;

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  return /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_MobileMenuItemModule.default.root, (_cn = {}, _cn[_MobileMenuItemModule.default["root_theme_" + theme]] = theme, _cn))
  }, /*#__PURE__*/_react.default.createElement(_MenuItemWithDropdown.default, {
    dropdownItems: dropdownItems,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_Link.default, {
    ref: ref,
    className: (0, _classnames.default)(_MobileMenuItemModule.default.link, (_cn2 = {}, _cn2[_MobileMenuItemModule.default.link_active] = active, _cn2[_MobileMenuItemModule.default.link_underlined] = underlined, _cn2)),
    "aria-current": active ? 'page' : undefined,
    path: path,
    onClick: onClick,
    dataQaType: dataQaType || 'uikit/navigation.mobileMenuItem',
    dataQaData: dataQaData
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _MobileMenuItemModule.default.title
  }, title), icon, count ? /*#__PURE__*/_react.default.createElement(_Badge.default, {
    count: count
  }) : null)), !count && newItem ? /*#__PURE__*/_react.default.createElement(_Badge.NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined
  }) : null);
});
MobileMenuItem.defaultProps = {
  underlined: true
};
var _default = MobileMenuItem;
exports.default = _default;
