// @platform-ui/notification/NotificationBase/NotificationBaseButtons.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _button = _interopRequireWildcard(require("@platform-ui/button"));

var _NotificationBaseButtonsModule = _interopRequireDefault(require("./NotificationBaseButtons.module.css"));

var NotificationBaseButtons = function NotificationBaseButtons(props) {
  var actionText = props.actionText,
      cancelText = props.cancelText,
      onActionConfirm = props.onActionConfirm,
      onCancel = props.onCancel;
  return actionText || cancelText ? /*#__PURE__*/_react.default.createElement("div", {
    className: _NotificationBaseButtonsModule.default.buttonGroup
  }, !!actionText && /*#__PURE__*/_react.default.createElement("span", {
    className: _NotificationBaseButtonsModule.default.actionButton
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    theme: "outlineDark",
    size: "s",
    onClick: onActionConfirm
  }, actionText)), !!cancelText && /*#__PURE__*/_react.default.createElement("span", {
    className: _NotificationBaseButtonsModule.default.actionButton
  }, /*#__PURE__*/_react.default.createElement(_button.ButtonLink, {
    theme: "flatDark",
    size: "s",
    pseudo: true,
    onClick: onCancel
  }, cancelText))) : null;
};

var _default = NotificationBaseButtons;
exports.default = _default;
