// @platform-ui/layout/Divider/Divider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _Box = _interopRequireDefault(require("../Box/Box"));

var _DividerModule = _interopRequireDefault(require("./Divider.module.css"));

var _excluded = ["color", "vertical"];
var Divider = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var color = _ref.color,
      vertical = _ref.vertical,
      props = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var width = vertical ? '1px' : '100%';
  var height = vertical ? undefined : '1px';
  var style = (0, _extends2.default)({
    backgroundColor: color
  }, vertical && {
    minHeight: '100%',
    alignSelf: 'stretch'
  });
  return /*#__PURE__*/_react.default.createElement(_Box.default, Object.assign({
    ref: ref,
    width: width,
    height: height,
    className: _DividerModule.default.root,
    style: style
  }, props));
});
var _default = Divider;
exports.default = _default;
