// @platform-ui/navigation/Logo/Logo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _NavigationContext = require("../NavigationContext");

var _LogoModule = _interopRequireDefault(require("./Logo.module.css"));

var Logo = function Logo(props) {
  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  var _props$lang = props.lang,
      lang = _props$lang === void 0 ? 'ru' : _props$lang;

  if (lang === 'ru') {
    var _cn;

    return /*#__PURE__*/_react.default.createElement("svg", Object.assign({
      xmlns: "http://www.w3.org/2000/svg",
      width: "104",
      height: "12",
      className: (0, _classnames.default)((_cn = {}, _cn[_LogoModule.default.root] = true, _cn[_LogoModule.default["root_theme_" + theme]] = theme, _cn))
    }, (0, _dataQa.default)(props)), /*#__PURE__*/_react.default.createElement("path", {
      fill: "currentColor",
      d: "M4.12 11.06h3V4.08h4.12V1.54H0v2.54h4.12v6.98zM20.34 1.54l-5 5.54V1.54h-2.9v9.52H15l5-5.54v5.54h2.9V1.54h-2.56zM28 7.52h4.66v3.54h2.9V1.54h-2.9v3.48H28V1.54h-2.9v9.52H28V7.52zM43.63 11.06c.66 0 1.26-.06 1.82-.22a3.29 3.29 0 002.48-3.3 3.2 3.2 0 00-2.18-3.18 6.18 6.18 0 00-2.12-.26h-2.98V1.54h-2.9v9.52h5.88zm-2.98-4.72h2.48c.4 0 .78 0 1.12.12.48.16.76.56.76 1.04 0 .5-.28.92-.74 1.12-.32.14-.64.14-1.14.14h-2.48V6.34zM52.33 7.46h.8l3.46 3.6h3.86L55.63 6.1l4.64-4.56h-3.64l-3.5 3.58h-.8V1.54h-2.9v9.52h2.9v-3.6z"
    }), /*#__PURE__*/_react.default.createElement("path", {
      fill: "currentColor",
      d: "M66.42 1.28c-3.9 0-6.24 2.2-6.24 5.04s2.34 5.04 6.24 5.04c3.88 0 6.22-2.2 6.22-5.04s-2.34-5.04-6.22-5.04zm0 7.76c-1.94 0-3.18-1.16-3.18-2.72 0-1.56 1.24-2.7 3.18-2.7 1.92 0 3.18 1.14 3.18 2.7 0 1.56-1.26 2.72-3.18 2.72zM82.3 1.96V1h-2.9v.96c-3.64.08-5.82 1.62-5.82 4.34 0 2.72 2.18 4.14 5.82 4.22v.96h2.9v-.96c3.66-.08 5.84-1.5 5.84-4.22 0-2.72-2.18-4.26-5.84-4.34zm-2.9 6.52c-2-.08-2.88-.9-2.88-2.18 0-1.28.88-2.2 2.88-2.28v4.46zm2.9 0V4.02c2.02.08 2.88 1 2.88 2.28 0 1.28-.86 2.1-2.88 2.18zM97.79 1.96V1h-2.9v.96c-3.64.08-5.82 1.62-5.82 4.34 0 2.72 2.18 4.14 5.82 4.22v.96h2.9v-.96c3.66-.08 5.84-1.5 5.84-4.22 0-2.72-2.18-4.26-5.84-4.34zm-2.9 6.52c-2-.08-2.88-.9-2.88-2.18 0-1.28.88-2.2 2.88-2.28v4.46zm2.9 0V4.02c2.02.08 2.88 1 2.88 2.28 0 1.28-.86 2.1-2.88 2.18z"
    }));
  }

  return /*#__PURE__*/_react.default.createElement("svg", {
    width: "75",
    height: "11",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/_react.default.createElement("path", {
    d: "M4.12 9.78h3V2.8h4.12V.26H0V2.8h4.12v6.98zM15.44 9.78V.26h-3v9.52h3zM20.316.26h-2.68v9.52h2.76V4.52l5.14 5.26h2.56V.26h-2.76v5.22L20.316.26zM33.192 6.18h.8l3.46 3.6h3.86l-4.82-4.96 4.64-4.56h-3.64l-3.5 3.58h-.8V.26h-2.9v9.52h2.9v-3.6z",
    fill: "currentColor"
  }), /*#__PURE__*/_react.default.createElement("path", {
    d: "M47.28 0c-3.9 0-6.24 2.2-6.24 5.04s2.34 5.04 6.24 5.04c3.88 0 6.22-2.2 6.22-5.04S51.16 0 47.28 0zm0 7.76c-1.94 0-3.18-1.16-3.18-2.72 0-1.56 1.24-2.7 3.18-2.7 1.92 0 3.18 1.14 3.18 2.7 0 1.56-1.26 2.72-3.18 2.72zM64.417 2.68V.26h-9.54v9.52h2.9V6.4h6.4v-2h-6.4V2.68h6.64zM75.14 2.68V.26H65.6v9.52h2.9V6.4h6.4v-2h-6.4V2.68h6.64z",
    fill: "currentColor"
  }));
};

var _default = Logo;
exports.default = _default;
