// @platform-ui/layout/esm/Box/Box.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["width", "height", "margin", "padding", "grow", "style", "tagName", "dataQaType", "dataQaData"];
import React, { forwardRef } from 'react';
import dataQa from '@platform-ui/dataQa';
import { getSpacing } from '../utils';
const Box = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    width,
    height,
    margin,
    padding,
    grow,
    style,
    tagName: TagName = 'div',
    dataQaType,
    dataQaData
  } = props,
        restProps = _objectWithoutPropertiesLoose(props, _excluded);

  return /*#__PURE__*/React.createElement(TagName, Object.assign({
    ref: ref,
    style: _extends({
      width: getSpacing(width),
      height: getSpacing(height),
      margin: getSpacing(margin),
      padding: getSpacing(padding),
      flexGrow: grow
    }, style)
  }, dataQa({
    dataQaType,
    dataQaData
  }), restProps));
});
export default Box;
