// @platform-ui/form/FormRow/FormRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _collapseBlock = _interopRequireDefault(require("@platform-ui/collapseBlock"));

var _generateId = require("@platform-ui/generateId");

var _overrideStyles = _interopRequireDefault(require("@platform-ui/overrideStyles"));

var _FormRowModule = _interopRequireDefault(require("./FormRow.module.css"));

var FormRow = function FormRow(props) {
  var _props$__overrideStyl, _cn, _cn2, _cn3, _cn4, _cn5, _cn6;

  var isWithMessage = function isWithMessage() {
    return !!(props.message || props.messageRight);
  };

  var isWithError = function isWithError() {
    return props.error && !!props.errorMessage;
  };

  var id = (0, _generateId.useId)();
  var errorId = "error_" + id;
  var messageId = "message_" + id;
  var styles = (0, _overrideStyles.default)(_FormRowModule.default, props == null ? void 0 : (_props$__overrideStyl = props.__overrideStyles) == null ? void 0 : _props$__overrideStyl.FormRow);
  var children = props.children,
      errorMessage = props.errorMessage,
      message = props.message,
      messageRight = props.messageRight,
      outerLabel = props.outerLabel,
      bigMessageOffset = props.bigMessageOffset,
      withoutOffset = props.withoutOffset,
      flexBasis = props.flexBasis,
      inFormGroup = props.inFormGroup,
      refCallback = props.refCallback,
      required = props.required,
      offset = props.offset,
      dataQaType = props.dataQaType,
      size = props.size;
  var withError = isWithError();
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_cn = {}, _cn[styles.root] = true, _cn[styles["root_offset-" + offset]] = offset, _cn[styles.root_withoutOffset] = withoutOffset, _cn[styles.root_inFormGroup] = inFormGroup, _cn)),
    style: {
      flexBasis: flexBasis
    },
    ref: refCallback,
    "data-qa-type": dataQaType
  }, outerLabel && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(styles.label, (_cn2 = {}, _cn2[styles["label_size-" + size]] = size, _cn2))
  }, outerLabel, required && '*'), typeof children === 'function' ? children({
    errorId: errorId,
    messageId: messageId
  }) : children, /*#__PURE__*/_react.default.createElement(_collapseBlock.default, {
    collapsed: !withError
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_cn3 = {}, _cn3[styles.errorBlock] = true, _cn3[styles.errorBlock_showed] = withError, _cn3)),
    id: errorId,
    "data-qa-type": dataQaType + ".errorBlock"
  }, errorMessage)), isWithMessage() && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_cn4 = {}, _cn4[styles.messageBlock] = true, _cn4[styles.messageBlock_withBigOffset] = bigMessageOffset, _cn4)),
    id: messageId,
    "data-qa-type": dataQaType + ".messageBlock"
  }, message && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_cn5 = {}, _cn5[styles.message] = true, _cn5[styles.message_left] = true, _cn5[styles.message_halfRow] = messageRight, _cn5))
  }, message), messageRight && /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)((_cn6 = {}, _cn6[styles.message] = true, _cn6[styles.message_right] = true, _cn6[styles.message_halfRow] = message, _cn6))
  }, messageRight)));
};

FormRow.defaultProps = {
  dataQaType: 'uikit/formRow',
  offset: 20,
  size: 'l'
};
var _default = FormRow;
exports.default = _default;
