// @platform-ui/notification/esm/NotificationFullScreen/NotificationFullScreen.js
import React from 'react';
import dataQa from '@platform-ui/dataQa';
import { Int24Attention as AttentionIcon, Int24Success as SuccessIcon } from '@platform-ui/iconsPack';
import Link from '@platform-ui/link';
import styles from './NotificationFullScreen.module.css';
const iconInfo = /*#__PURE__*/React.createElement(AttentionIcon, null);
const iconSuccess = /*#__PURE__*/React.createElement(SuccessIcon, null);

const NotificationFullScreen = props => {
  const {
    type,
    withIcon,
    icon,
    children,
    linkText,
    linkHref,
    linkPseudo
  } = props;
  return /*#__PURE__*/React.createElement("div", Object.assign({}, dataQa(props), {
    className: `${styles.root} ${styles[`root_${type}`]}`
  }), withIcon && /*#__PURE__*/React.createElement("div", {
    className: styles.icon
  }, !icon && (type === 'success' && iconSuccess || iconInfo), icon), /*#__PURE__*/React.createElement("div", {
    className: styles.content
  }, children, linkText && /*#__PURE__*/React.createElement("div", {
    className: styles.link
  }, /*#__PURE__*/React.createElement(Link, {
    href: linkHref,
    pseudo: linkPseudo
  }, linkText))));
}; // TODO: убрать в мажорном обновлении


NotificationFullScreen.defaultProps = {
  dataQaType: 'uikit/notificationFullScreen',
  type: 'info',
  withIcon: true
};
export default NotificationFullScreen;
