// @platform-ui/navigation/User/Companies/CompaniesList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _scroll = _interopRequireDefault(require("@platform-ui/scroll"));

var _classnames = _interopRequireDefault(require("classnames"));

var _index = require("./index");

var _CompaniesListModule = _interopRequireDefault(require("./CompaniesList.module.css"));

var ListWrapper = function ListWrapper(_ref) {
  var children = _ref.children,
      withScroll = _ref.withScroll;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, withScroll ? /*#__PURE__*/_react.default.createElement(_scroll.default, null, children) : children);
};

var CompaniesList = function CompaniesList(_ref2) {
  var _cn;

  var companies = _ref2.companies,
      profileLink = _ref2.profileLink,
      onClick = _ref2.onClick,
      mobile = _ref2.mobile;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, companies && companies.length ? /*#__PURE__*/_react.default.createElement(ListWrapper, {
    withScroll: !mobile
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_CompaniesListModule.default.companies, (_cn = {}, _cn[_CompaniesListModule.default.companies_mobile] = mobile, _cn)),
    style: !mobile && companies && companies.length > 6 ? {
      maxHeight: 280
    } : {}
  }, profileLink && /*#__PURE__*/_react.default.createElement(_index.CompaniesItem, Object.assign({
    onClick: onClick
  }, profileLink, {
    mobile: true
  })), /*#__PURE__*/_react.default.createElement(_index.Companies, {
    companies: companies,
    onClick: onClick,
    mobile: true
  }))) : null);
};

var _default = CompaniesList;
exports.default = _default;
