// @platform-ui/navigation/esm/User/UserMobile/UserMobileRoot.js
import React, { useState } from 'react';
import IconMenu from '../Icons/Menu';
import MobilePopup from '../../MobilePopup';
import styles from './UserMobileRoot.module.css';

const UserMobileRoot = props => {
  const {
    children: profileComponent
  } = props;
  const [menuVisible, setMenuVisible] = useState(false);
  return /*#__PURE__*/React.createElement("div", {
    className: styles.root
  }, /*#__PURE__*/React.createElement("div", {
    role: "button",
    className: styles.icon,
    onClick: () => {
      setMenuVisible(true);
    }
  }, /*#__PURE__*/React.createElement(IconMenu, null)), /*#__PURE__*/React.createElement(MobilePopup, {
    active: menuVisible,
    align: "right",
    onClose: () => {
      setMenuVisible(false);
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: styles.content
  }, /*#__PURE__*/React.createElement(React.Fragment, null, typeof profileComponent === 'function' ? profileComponent({
    closeProfilePopover: () => setMenuVisible(false)
  }) : profileComponent))));
};

export default UserMobileRoot;
