// @platform-ui/navigation/esm/User/UserMobile/UserMobile.js
import React from 'react';
import Logout from '../Logout';
import Profile from '../Profile';
import CompaniesList from '../Companies';
import { Divider } from '../../index';
import UserMobileRoot from './UserMobileRoot';
import UserMobileSpacer from './UserMobileSpacer';

const UserMobile = props => {
  var _user$companies;

  const {
    user,
    logoutLink,
    onClick
  } = props;
  return /*#__PURE__*/React.createElement(UserMobileRoot, null, /*#__PURE__*/React.createElement(Profile, {
    mobile: true,
    avatarUrl: user.avatarUrl,
    fullName: user.fullName,
    premium: user.premium,
    links: user.profileSections,
    onClick: onClick,
    badgeSubscriptions: user.badgeSubscriptions
  }), (_user$companies = user.companies) != null && _user$companies.length ? /*#__PURE__*/React.createElement(Divider, null) : null, /*#__PURE__*/React.createElement(CompaniesList, {
    profileLink: user.profileLink,
    onClick: onClick,
    companies: user.companies,
    mobile: true
  }), /*#__PURE__*/React.createElement(UserMobileSpacer, null), /*#__PURE__*/React.createElement(Logout, Object.assign({
    onClick: onClick
  }, logoutLink, {
    mobile: true
  })));
};

export default UserMobile;
