// @platform-ui/popup/templates/Error/Error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _reactUse = require("react-use");

var _classnames = _interopRequireDefault(require("classnames"));

var _button = _interopRequireWildcard(require("@platform-ui/button"));

var _Alert = _interopRequireDefault(require("@platform-ui/iconsPack/markers/ui/Alert"));

var _mediaBreakpoints = _interopRequireDefault(require("@platform-ui/constants/mediaBreakpoints"));

var _Popup = _interopRequireDefault(require("../../Popup"));

var _ErrorModule = _interopRequireDefault(require("./Error.module.css"));

var _excluded = ["title", "description", "approve", "decline", "onClose", "size"];

function Error(_ref) {
  var _cn, _cn2;

  var title = _ref.title,
      description = _ref.description,
      approve = _ref.approve,
      decline = _ref.decline,
      onClose = _ref.onClose,
      _ref$size = _ref.size,
      size = _ref$size === void 0 ? 'm' : _ref$size,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var tabletS = _mediaBreakpoints.default.tabletS;
  var mediaTablet = (0, _reactUse.useMedia)(tabletS, true);
  var handleClose = (0, _react.useCallback)(function (event, params) {
    onClose && onClose(event, {
      value: params && typeof params.value !== 'undefined' ? params.value : undefined
    });
  }, [onClose]);
  return /*#__PURE__*/_react.default.createElement(_Popup.default, Object.assign({}, otherProps, {
    size: size,
    mobileSize: "fullscreen",
    onClose: handleClose
  }), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      display: 'flex',
      flexDirection: 'column',
      minHeight: mediaTablet ? 'auto' : 'calc(100vh - 64px)'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_ErrorModule.default.icon, (_cn = {}, _cn[_ErrorModule.default["icon_desktop-" + size]] = mediaTablet, _cn[_ErrorModule.default.icon_mobile] = !mediaTablet, _cn))
  }, /*#__PURE__*/_react.default.createElement(_Alert.default, {
    theme: "red",
    withContainer: true,
    iconSize: !mediaTablet || size === 'm' ? 96 : 64
  })), /*#__PURE__*/_react.default.createElement("h2", {
    className: (0, _classnames.default)(_ErrorModule.default.title, (_cn2 = {}, _cn2[_ErrorModule.default.title_desktop] = mediaTablet, _cn2[_ErrorModule.default.title_mobile] = !mediaTablet, _cn2))
  }, title), /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorModule.default.content
  }, description), /*#__PURE__*/_react.default.createElement("div", {
    className: _ErrorModule.default.buttons
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    size: size === 'm' && mediaTablet ? 'l' : 'm',
    wide: !mediaTablet,
    onClick: (0, _react.useCallback)(function (event) {
      handleClose(event, {
        value: true
      });
    }, [handleClose])
  }, approve), decline && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
    className: _ErrorModule.default.separator
  }), /*#__PURE__*/_react.default.createElement(_button.ButtonLink, {
    size: size === 'm' && mediaTablet ? 'l' : 'm',
    wide: !mediaTablet,
    onClick: function onClick(event) {
      handleClose(event, {
        value: false
      });
    }
  }, decline)))));
}

var _default = Error;
exports.default = _default;
