// @platform-ui/scroll/ScrollContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.ScrollContent = void 0;

var _react = _interopRequireWildcard(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _domUtils = require("@platform-ui/dom-utils");

var _classnames = _interopRequireDefault(require("classnames"));

var _ScrollModule = _interopRequireDefault(require("./Scroll.module.css"));

var _utils = require("./utils");

var ScrollContent = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var _cn;

  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/scroll' : _props$dataQaType,
      scrollY = props.scrollY,
      scrollX = props.scrollX,
      onScroll = props.onScroll,
      onWheel = props.onWheel,
      id = props.id,
      children = props.children;
  var width = (0, _domUtils.useScrollbarWidth)();
  return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType + ".content"
  }), {
    ref: ref,
    className: (0, _classnames.default)((_cn = {}, _cn[_ScrollModule.default.content] = true, _cn[_ScrollModule.default.content_scrollY] = scrollY, _cn[_ScrollModule.default.content_scrollX] = scrollX, _cn)),
    id: id,
    style: (0, _utils.windowStyles)(scrollY, scrollX, width),
    onScroll: onScroll,
    onWheel: onWheel
  }), children);
});
exports.ScrollContent = ScrollContent;
