// @platform-ui/scroll/ScrollBar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dragHOC = _interopRequireDefault(require("@platform-ui/dragHOC"));

var _ScrollModule = _interopRequireDefault(require("./Scroll.module.css"));

var _class, _class2;

var noop = function noop() {};

var mainProperty = {
  vertical: 'height',
  horizontal: 'width'
};

var ScrollBar = (0, _dragHOC.default)(_class = (_class2 = /*#__PURE__*/function (_PureComponent) {
  (0, _inheritsLoose2.default)(ScrollBar, _PureComponent);

  function ScrollBar() {
    var _this;

    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    _this = _PureComponent.call.apply(_PureComponent, [this].concat(_args)) || this;

    _this.onDragStart = function () {
      var _this$props;

      return (_this$props = _this.props).handleDragStart.apply(_this$props, arguments);
    };

    _this.onDragProcess = function () {
      var _this$props2;

      return (_this$props2 = _this.props).handleDragProcess.apply(_this$props2, arguments);
    };

    _this.onDragEnd = function () {
      var _this$props3;

      return (_this$props3 = _this.props).handleDragEnd.apply(_this$props3, arguments);
    };

    _this.handleMouseDown = function (event) {
      var _this$props4 = _this.props,
          onMouseDown = _this$props4.onMouseDown,
          onDrag = _this$props4.onDrag,
          direction = _this$props4.direction;
      onMouseDown(event, direction);
      onDrag(event);
    };

    return _this;
  }

  var _proto = ScrollBar.prototype;

  _proto.render = function render() {
    var _cn, _ref;

    var _this$props5 = this.props,
        visible = _this$props5.visible,
        barRef = _this$props5.barRef,
        size = _this$props5.size,
        theme = _this$props5.theme,
        dragging = _this$props5.dragging,
        direction = _this$props5.direction;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)((_cn = {}, _cn[_ScrollModule.default.scrollBar] = true, _cn[_ScrollModule.default.scrollBar_visible] = visible, _cn[_ScrollModule.default.scrollBar_dragging] = dragging, _cn[_ScrollModule.default["scrollBar_direction_" + direction]] = true, _cn[_ScrollModule.default["scrollBar_theme_" + theme]] = true, _cn))
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: _ScrollModule.default.scrollCaret,
      ref: barRef,
      style: (_ref = {}, _ref[mainProperty[direction]] = size, _ref),
      onMouseDown: this.handleMouseDown
    }));
  };

  return ScrollBar;
}(_react.PureComponent), _class2.defaultProps = {
  visible: false,
  dragging: false,
  size: 0,
  theme: 'default',
  onDrag: noop,
  onMouseDown: noop,
  handleDragStart: noop,
  handleDragProcess: noop,
  handleDragEnd: noop
}, _class2)) || _class;

var _default = ScrollBar;
exports.default = _default;
