// @platform-ui/island/Island.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var corners = _interopRequireWildcard(require("@platform-ui/flatCorners"));

var _collapseBlock = _interopRequireDefault(require("@platform-ui/collapseBlock"));

var _link = _interopRequireDefault(require("@platform-ui/link"));

var _Island = _interopRequireDefault(require("./Island.context"));

var _IslandContent = _interopRequireDefault(require("./IslandContent/IslandContent"));

var _IslandFooterButton = _interopRequireDefault(require("./IslandFooterButton/IslandFooterButton"));

var _IslandModule = _interopRequireDefault(require("./Island.module.css"));

var _types = require("./types");

var _excluded = ["size", "theme", "inGroup", "vertical", "clickable", "fillHeight", "flatCorners", "children", "footer", "buttonText", "buttonHref", "buttonDisabled", "dataQaType", "onButtonClick", "onClick", "collapsible", "defaultCollapsed", "expandButtonText", "collapseButtonText", "collapseMinHeight", "collapseDuration"];

var Island = function Island(props) {
  var _cn;

  var size = props.size,
      theme = props.theme,
      inGroup = props.inGroup,
      vertical = props.vertical,
      clickable = props.clickable,
      fillHeight = props.fillHeight,
      flatCorners = props.flatCorners,
      children = props.children,
      footer = props.footer,
      buttonText = props.buttonText,
      buttonHref = props.buttonHref,
      buttonDisabled = props.buttonDisabled,
      dataQaType = props.dataQaType,
      onButtonClick = props.onButtonClick,
      onClick = props.onClick,
      collapsible = props.collapsible,
      defaultCollapsed = props.defaultCollapsed,
      expandButtonText = props.expandButtonText,
      collapseButtonText = props.collapseButtonText,
      collapseMinHeight = props.collapseMinHeight,
      collapseDuration = props.collapseDuration,
      contentProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  var _useState = (0, _react.useState)(defaultCollapsed),
      collapsed = _useState[0],
      setCollapsed = _useState[1];

  var groupDirection = inGroup && (vertical ? 'vertical' : 'horizontal');

  var handleExpandClick = function handleExpandClick() {
    setCollapsed(!collapsed);
  };

  return /*#__PURE__*/_react.default.createElement(_Island.default.Provider, {
    value: size
  }, /*#__PURE__*/_react.default.createElement("div", {
    onClick: clickable ? onClick : undefined,
    className: (0, _classnames.default)((_cn = {}, _cn[_IslandModule.default.island] = true, _cn[_IslandModule.default.island_clickable] = clickable, _cn[_IslandModule.default.island_fillHeight] = fillHeight, _cn[_IslandModule.default["island_size_" + size]] = true, _cn[_IslandModule.default["island_theme_" + theme]] = true, _cn[_IslandModule.default.island_inGroup] = inGroup, _cn[_IslandModule.default["island_inGroup_" + groupDirection]] = groupDirection, _cn[_IslandModule.default.island_collapsed] = collapsible && collapsed, _cn)),
    "data-qa-type": dataQaType,
    style: corners.style(flatCorners)
  }, collapsible ? /*#__PURE__*/_react.default.createElement("div", {
    className: _IslandModule.default.collapseBlockContainer,
    style: {
      animationDuration: collapseDuration + "ms"
    }
  }, /*#__PURE__*/_react.default.createElement(_collapseBlock.default, {
    collapsed: collapsed,
    duration: collapseDuration,
    removeCollapsedContent: false,
    minHeight: collapseMinHeight
  }, children || /*#__PURE__*/_react.default.createElement(_IslandContent.default, contentProps))) : children || /*#__PURE__*/_react.default.createElement(_IslandContent.default, contentProps), !collapsible ? null : /*#__PURE__*/_react.default.createElement("div", {
    className: _IslandModule.default.expandControls
  }, /*#__PURE__*/_react.default.createElement(_link.default, {
    onClick: handleExpandClick
  }, collapsed ? expandButtonText : collapseButtonText)), !!(footer || buttonText) && /*#__PURE__*/_react.default.createElement("footer", {
    className: _IslandModule.default.footer
  }, footer, buttonText && /*#__PURE__*/_react.default.createElement(_IslandFooterButton.default, {
    href: buttonHref,
    onClick: onButtonClick,
    disabled: buttonDisabled
  }, buttonText))));
};

Island.defaultProps = _types.DEFAULT_PROPS;
var _default = Island;
exports.default = _default;
