// @platform-ui/inputTag/TagInput/TagEditInputAutocomplete.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _TagInputStandalone = _interopRequireDefault(require("./TagInputStandalone"));

var _TagsContext = require("../TagsContext");

var TagEditInputAutocomplete = function TagEditInputAutocomplete() {
  var context = (0, _TagsContext.useTagsContextAutocomplete)();
  var value = context.value,
      opened = context.opened,
      setOpened = context.setOpened,
      highlightedOption = context.highlightedOption,
      setInputValue = context.setInputValue,
      setFocusedIndex = context.setFocusedIndex,
      setEditing = context.setEditing,
      onReplace = context.onReplace,
      _onBlur = context.onBlur,
      onHighlightPrev = context.onHighlightPrev,
      onHighlightNext = context.onHighlightNext,
      acceptTag = context.acceptTag,
      dataQaType = context.dataQaType;
  var inputValue = context.inputValue;
  var index = context.focusedIndex;
  return /*#__PURE__*/_react.default.createElement(_TagInputStandalone.default, {
    dataQaType: dataQaType + ".editInputAutocomplete",
    value: inputValue,
    focused: true,
    onBlur: function onBlur(event) {
      var tag = acceptTag(highlightedOption || inputValue);

      if (tag !== null) {
        onReplace(event, {
          index: index,
          tag: tag
        });
      }

      _onBlur(event);
    },
    onCancel: function onCancel() {
      if (opened) {
        setOpened(false);
        return;
      }

      setEditing(false);
      setInputValue('');
    },
    onChange: function onChange(event, params) {
      setInputValue(params.value);
    },
    onApply: function onApply(event) {
      var tag = acceptTag(highlightedOption || inputValue);

      if (tag !== null) {
        onReplace(event, {
          index: index,
          tag: tag
        });
      }
    },
    onExitBackward: function onExitBackward(event) {
      if (index === 0) {
        return;
      }

      var tag = acceptTag(highlightedOption || inputValue);

      if (tag !== null) {
        onReplace(event, {
          index: index,
          tag: tag
        });
      } else {
        setEditing(false);
        setInputValue('');
      }

      setFocusedIndex(index - 1);
    },
    onExitForward: function onExitForward(event) {
      var tag = acceptTag(highlightedOption || inputValue);
      setFocusedIndex(index === value.length - 1 ? 'input' : index + 1);

      if (tag !== null) {
        onReplace(event, {
          index: index,
          tag: tag
        });
      } else {
        setEditing(false);
        setInputValue('');
      }
    },
    onOptionNext: onHighlightNext,
    onOptionPrev: onHighlightPrev
  });
};

var _default = TagEditInputAutocomplete;
exports.default = _default;
