// @platform-ui/toggle/Toggle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = exports.ToggleStateless = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _iconsPack = require("@platform-ui/iconsPack");

var _checkable = _interopRequireDefault(require("@platform-ui/checkable"));

var _stateUtils = require("@platform-ui/stateUtils");

var _ToggleModule = _interopRequireDefault(require("./Toggle.module.css"));

var _excluded = ["dataQaType", "size", "theme", "loading", "withIcons", "inputAttrs", "color"];

var ToggleStateless = function ToggleStateless(props) {
  var _cn;

  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/toggle' : _props$dataQaType,
      _props$size = props.size,
      size = _props$size === void 0 ? 'l' : _props$size,
      _props$theme = props.theme,
      theme = _props$theme === void 0 ? 'default' : _props$theme,
      loading = props.loading,
      _props$withIcons = props.withIcons,
      withIcons = _props$withIcons === void 0 ? true : _props$withIcons,
      inputAttrs = props.inputAttrs,
      color = props.color,
      inputProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  return /*#__PURE__*/_react.default.createElement("span", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType
  }), {
    className: (0, _classnames.default)(_ToggleModule.default.root, _ToggleModule.default["root_size_" + size], _ToggleModule.default["root_theme_" + theme], (_cn = {}, _cn[_ToggleModule.default.root_loading] = loading, _cn))
  }), /*#__PURE__*/_react.default.createElement(_checkable.default, Object.assign({}, inputAttrs, inputProps, {
    dataQaType: dataQaType + ".input",
    type: "checkbox",
    className: _ToggleModule.default.input,
    style: inputProps.checked && color ? {
      backgroundColor: color
    } : undefined
  })), withIcons && /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_ToggleModule.default.icon, _ToggleModule.default.icon_type_checked)
  }, size === 's' ? /*#__PURE__*/_react.default.createElement(_iconsPack.Int16Tip, null) : /*#__PURE__*/_react.default.createElement(_iconsPack.Int24Tip, null)), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_ToggleModule.default.icon, _ToggleModule.default.icon_type_unchecked)
  })), /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_ToggleModule.default.icon, _ToggleModule.default.icon_type_loading)
  }));
};

exports.ToggleStateless = ToggleStateless;
var Toggle = (0, _stateUtils.withCheckableInputState)(ToggleStateless);
var _default = Toggle;
exports.default = _default;
