// @platform-ui/carousel/esm/CarouselControlsGlobal.js
import React from 'react';
import cn from 'classnames';
import dataQa from '@platform-ui/dataQa';
import { Int24ArrowDown as ArrowIcon } from '@platform-ui/iconsPack';
import s from './Carousel.module.css';

const Controls = props => {
  const {
    dataQaType,
    carouselId,
    onNextClick: handleNextClick,
    onPrevClick: handlePrevClick,
    prevAriaLabel = 'Previous',
    nextAriaLabel = 'Next',
    nextDisabled,
    prevDisabled
  } = props;
  return /*#__PURE__*/React.createElement("div", Object.assign({
    className: s.controlContainer,
    "aria-controls": carouselId
  }, dataQa(props)), /*#__PURE__*/React.createElement("div", Object.assign({
    className: cn({
      [s.button]: true,
      [s.button_left]: s.button_left,
      [s.button_disabled]: prevDisabled
    }),
    "aria-label": prevAriaLabel,
    onClick: !prevDisabled ? handlePrevClick : undefined
  }, dataQa({
    dataQaType: `${dataQaType}.back`
  })), /*#__PURE__*/React.createElement(ArrowIcon, {
    rotate: 90
  })), /*#__PURE__*/React.createElement("div", Object.assign({
    className: cn({
      [s.button]: true,
      [s.button_right]: s.button_right,
      [s.button_disabled]: nextDisabled
    }),
    "aria-label": nextAriaLabel,
    onClick: !nextDisabled ? handleNextClick : undefined
  }, dataQa({
    dataQaType: `${dataQaType}.forward`
  })), /*#__PURE__*/React.createElement(ArrowIcon, {
    rotate: 270
  })));
};

export default Controls;
