// @platform-ui/notification/NotificationStack/NotificationStack.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _NotificationToast = _interopRequireDefault(require("../NotificationToast/NotificationToast"));

var _NotificationStackModule = _interopRequireDefault(require("./NotificationStack.module.css"));

var _excluded = ["dataQaType", "paddingBottom", "notifications", "offsetTop", "offsetRight", "width", "zIndex", "onHide", "onClose", "onActionConfirm", "onCancel"],
    _excluded2 = ["notificationComponent"];

var NotificationStack = function NotificationStack(props) {
  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/NotificationStack' : _props$dataQaType,
      _props$paddingBottom = props.paddingBottom,
      paddingBottom = _props$paddingBottom === void 0 ? 16 : _props$paddingBottom,
      notifications = props.notifications,
      offsetTop = props.offsetTop,
      offsetRight = props.offsetRight,
      width = props.width,
      zIndex = props.zIndex,
      onHide = props.onHide,
      onClose = props.onClose,
      onActionConfirm = props.onActionConfirm,
      onCancel = props.onCancel,
      baseProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-qa-type": dataQaType,
    className: _NotificationStackModule.default.root,
    style: {
      top: offsetTop,
      right: offsetRight,
      width: width,
      zIndex: zIndex
    }
  }, Array.isArray(notifications) && notifications.map(function (notification, index) {
    var _notification$notific = notification.notificationComponent,
        NotificationComponent = _notification$notific === void 0 ? _NotificationToast.default : _notification$notific,
        notificationProps = (0, _objectWithoutPropertiesLoose2.default)(notification, _excluded2);
    return /*#__PURE__*/_react.default.createElement(NotificationComponent, Object.assign({
      key: notificationProps.id || index,
      onHide: onHide ? function (event) {
        return onHide(event, notificationProps);
      } : undefined,
      onClose: onClose ? function (event) {
        return onClose(event, notificationProps);
      } : undefined,
      onActionConfirm: onActionConfirm ? function (event) {
        return onActionConfirm(event, notificationProps);
      } : undefined,
      onCancel: onCancel ? function (event) {
        return onCancel(event, notificationProps);
      } : undefined,
      paddingBottom: paddingBottom
    }, baseProps, notificationProps));
  }));
}; // TODO: убрать в мажорном обновлении


NotificationStack.defaultProps = (0, _extends2.default)({}, _NotificationToast.default.defaultProps, {
  dataQaType: 'uikit/NotificationStack',
  // TODO: переименовать проп так, чтобы было понятно что именно он делает - задаёт отступ между нотификациями
  paddingBottom: 16
});
var _default = NotificationStack;
exports.default = _default;
