// @platform-ui/navigation/MenuItemWithDropdown/MenuItemWithDropdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _contextMenu = _interopRequireDefault(require("@platform-ui/contextMenu"));

var _deviceInfo = require("@platform-ui/device-info");

var _utils = require("../utils");

var _MenuItemWithDropdown = _interopRequireDefault(require("./MenuItemWithDropdown.css"));

function MenuItemWithDropdown(_ref) {
  var children = _ref.children,
      _ref$dropdownItems = _ref.dropdownItems,
      dropdownItems = _ref$dropdownItems === void 0 ? [] : _ref$dropdownItems,
      onClick = _ref.onClick;

  var _useState = (0, _react.useState)(false),
      touchDevice = _useState[0],
      setTouchDevice = _useState[1];

  var memoizedItems = (0, _react.useMemo)(function () {
    return dropdownItems == null ? void 0 : dropdownItems.map(function (item) {
      return (0, _utils.menuItemToContextMenu)(item, onClick);
    });
  }, [dropdownItems, onClick]);
  (0, _react.useEffect)(function () {
    setTouchDevice((0, _deviceInfo.isTouchDevice)());
  }, []);
  return memoizedItems.length ? /*#__PURE__*/_react.default.createElement(_contextMenu.default, {
    items: memoizedItems,
    showOnHover: true,
    align: "start",
    itemSize: "xs"
  }, children, touchDevice ? /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemWithDropdown.default.holder,
    tabIndex: -1,
    role: "presentation"
  }) : null) : children;
}

var _default = MenuItemWithDropdown;
exports.default = _default;
