// @platform-ui/scroll/esm/ScrollBar.js
var _class, _class2;

import React, { PureComponent } from 'react';
import cn from 'classnames'; // @ts-ignore

import dragHOC from '@platform-ui/dragHOC';
import styles from './Scroll.module.css';

const noop = () => {};

const mainProperty = {
  vertical: 'height',
  horizontal: 'width'
};

let ScrollBar = dragHOC(_class = (_class2 = class ScrollBar extends PureComponent {
  constructor(..._args) {
    super(..._args);

    this.onDragStart = (...args) => this.props.handleDragStart(...args);

    this.onDragProcess = (...args) => this.props.handleDragProcess(...args);

    this.onDragEnd = (...args) => this.props.handleDragEnd(...args);

    this.handleMouseDown = event => {
      const {
        onMouseDown,
        onDrag,
        direction
      } = this.props;
      onMouseDown(event, direction);
      onDrag(event);
    };
  }

  render() {
    const {
      visible,
      barRef,
      size,
      theme,
      dragging,
      direction
    } = this.props;
    return /*#__PURE__*/React.createElement("div", {
      className: cn({
        [styles.scrollBar]: true,
        [styles.scrollBar_visible]: visible,
        [styles.scrollBar_dragging]: dragging,
        [styles[`scrollBar_direction_${direction}`]]: true,
        [styles[`scrollBar_theme_${theme}`]]: true
      })
    }, /*#__PURE__*/React.createElement("div", {
      className: styles.scrollCaret,
      ref: barRef,
      style: {
        [mainProperty[direction]]: size
      },
      onMouseDown: this.handleMouseDown
    }));
  }

}, _class2.defaultProps = {
  visible: false,
  dragging: false,
  size: 0,
  theme: 'default',
  onDrag: noop,
  onMouseDown: noop,
  handleDragStart: noop,
  handleDragProcess: noop,
  handleDragEnd: noop
}, _class2)) || _class;

export default ScrollBar;
