// @platform-ui/pageHeader/esm/EditableRow/Controls.js
import React from 'react';
import cn from 'classnames';
import Button from '@platform-ui/button';
import TipIcon from '@platform-ui/iconsPack/interface/24/Tip';
import CloseIcon from '@platform-ui/iconsPack/interface/24/Close';
import styles from './Controls.module.css';

const Controls = props => {
  const {
    onSave,
    onCancel,
    mobile,
    saveButtonLabel = 'Сохранить',
    cancelButtonLabel = 'Отменить'
  } = props;
  return /*#__PURE__*/React.createElement("div", {
    className: cn(styles.root, {
      [styles.root_mobile]: mobile
    })
  }, /*#__PURE__*/React.createElement(Button, {
    size: "s",
    theme: "flat",
    onClick: onCancel,
    icon: mobile ? CloseIcon : null,
    ariaLabel: mobile ? cancelButtonLabel : undefined
  }, mobile ? null : cancelButtonLabel), /*#__PURE__*/React.createElement(Button, {
    size: "s",
    theme: mobile ? 'flat' : 'secondary',
    onClick: onSave,
    icon: mobile ? TipIcon : null,
    ariaLabel: mobile ? saveButtonLabel : undefined
  }, mobile ? null : saveButtonLabel));
};

export default Controls;
