// @platform-ui/button/ButtonLink.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _omit = _interopRequireDefault(require("@tinkoff/utils/object/omit"));

var _link = _interopRequireDefault(require("@platform-ui/link"));

var _Icon = _interopRequireDefault(require("./Icon"));

var _Button = _interopRequireDefault(require("./Button"));

var _useButtonStyles = _interopRequireDefault(require("./useButtonStyles"));

var dropKeysFromObject = function dropKeysFromObject(o, keys) {
  var copy = (0, _extends2.default)({}, o);
  keys.forEach(function (key) {
    return delete copy[key];
  });
  return copy;
};

var ButtonLink = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var dataQaData = props.dataQaData,
      dataQaType = props.dataQaType,
      type = props.type,
      size = props.size,
      icon = props.icon,
      children = props.children,
      onClick = props.onClick,
      onMouseLeave = props.onMouseLeave,
      onMouseOver = props.onMouseOver,
      innerRef = props.innerRef,
      href = props.href,
      progressiveFill = props.progressiveFill,
      pseudo = props.pseudo,
      disabled = props.disabled,
      name = props.name,
      theme = props.theme,
      target = props.target,
      rel = props.rel,
      download = props.download,
      tabIndex = props.tabIndex;
  var linkProps = {
    onClick: onClick,
    onMouseLeave: onMouseLeave,
    onMouseOver: onMouseOver,
    type: type,
    href: href,
    pseudo: pseudo,
    disabled: disabled,
    name: name,
    theme: theme,
    target: target,
    rel: rel,
    download: download,
    tabIndex: tabIndex
  };
  var styleProps = dropKeysFromObject(props, ['disabled', 'theme', 'pseudo', 'href']);
  var buttonStyles = (0, _useButtonStyles.default)((0, _extends2.default)({}, styleProps, {
    isButtonLink: true
  }));
  return /*#__PURE__*/_react.default.createElement("div", {
    className: buttonStyles.button.className,
    style: buttonStyles.button.style,
    "data-size": size,
    "data-qa-data": dataQaData,
    "data-qa-type": dataQaType
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: buttonStyles.buttonWrapper.className
  }, /*#__PURE__*/_react.default.createElement(_link.default, Object.assign({}, linkProps, {
    ref: ref,
    innerRef: innerRef
  }), progressiveFill ? /*#__PURE__*/_react.default.createElement("span", {
    className: buttonStyles.progressiveFillContainer.className
  }, /*#__PURE__*/_react.default.createElement("span", {
    "data-qa-type": dataQaType + ".progressiveFill",
    className: buttonStyles.progressiveFill.className,
    style: buttonStyles.progressiveFill.style
  })) : null, icon && /*#__PURE__*/_react.default.createElement("span", {
    className: buttonStyles.icon.className,
    "data-qa-type": dataQaType + ".icon"
  }, /*#__PURE__*/_react.default.createElement(_Icon.default, {
    icon: icon
  })), children && /*#__PURE__*/_react.default.createElement("span", {
    className: buttonStyles.content.className,
    "data-qa-type": dataQaType + ".content"
  }, children))));
});
ButtonLink.defaultProps = (0, _extends2.default)({}, (0, _omit.default)(['theme'], _Button.default.defaultProps)); // TODO: разобраться как подружить сборку TS и displayName

ButtonLink.displayName = 'ButtonLink';
var _default = ButtonLink;
exports.default = _default;
