// @platform-ui/navigation/esm/User/Companies/CompaniesList.js
import React from 'react';
import Scroll from '@platform-ui/scroll';
import cn from 'classnames';
import { Companies, CompaniesItem } from './index';
import styles from './CompaniesList.module.css';

const ListWrapper = ({
  children,
  withScroll
}) => {
  return /*#__PURE__*/React.createElement(React.Fragment, null, withScroll ? /*#__PURE__*/React.createElement(Scroll, null, children) : children);
};

const CompaniesList = ({
  companies,
  profileLink,
  onClick,
  mobile
}) => {
  return /*#__PURE__*/React.createElement(React.Fragment, null, companies && companies.length ? /*#__PURE__*/React.createElement(ListWrapper, {
    withScroll: !mobile
  }, /*#__PURE__*/React.createElement("div", {
    className: cn(styles.companies, {
      [styles.companies_mobile]: mobile
    }),
    style: !mobile && companies && companies.length > 6 ? {
      maxHeight: 280
    } : {}
  }, profileLink && /*#__PURE__*/React.createElement(CompaniesItem, Object.assign({
    onClick: onClick
  }, profileLink, {
    mobile: true
  })), /*#__PURE__*/React.createElement(Companies, {
    companies: companies,
    onClick: onClick,
    mobile: true
  }))) : null);
};

export default CompaniesList;
