// @platform-ui/navigation/Menu/MenuItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Link = _interopRequireDefault(require("../Link"));

var _Badge = _interopRequireWildcard(require("../Badge"));

var _MoreButton = _interopRequireDefault(require("./MoreButton"));

var _MenuItemWithDropdown = _interopRequireDefault(require("../MenuItemWithDropdown/MenuItemWithDropdown"));

var _MenuItemModule = _interopRequireDefault(require("./MenuItem.module.css"));

var _NavigationContext = require("../NavigationContext");

var MenuItem = function MenuItem(props) {
  var _cn;

  var title = props.title,
      active = props.active,
      underlined = props.underlined,
      path = props.path,
      icon = props.icon,
      count = props.count,
      contextItems = props.contextItems,
      onClick = props.onClick,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      hidden = props.hidden,
      newItem = props.newItem,
      dropdownItems = props.dropdownItems;

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  return /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemModule.default.root
  }, /*#__PURE__*/_react.default.createElement(_MenuItemWithDropdown.default, {
    dropdownItems: dropdownItems,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_Link.default, {
    className: (0, _classnames.default)(_MenuItemModule.default.link, (_cn = {}, _cn[_MenuItemModule.default.link_active] = active, _cn[_MenuItemModule.default["link_theme_" + theme]] = theme, _cn[_MenuItemModule.default.link_underlined] = underlined, _cn)),
    "aria-current": active ? 'page' : undefined,
    path: path,
    tabIndex: hidden ? -1 : undefined,
    onClick: onClick,
    dataQaType: dataQaType || 'uikit/navigation.menuItem',
    dataQaData: dataQaData
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemModule.default.title
  }, title), icon ? /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemModule.default.icon
  }, icon) : null, count ? /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemModule.default.badge
  }, /*#__PURE__*/_react.default.createElement(_Badge.default, {
    count: count
  })) : null), ' ')), contextItems ? /*#__PURE__*/_react.default.createElement("span", {
    className: _MenuItemModule.default.more
  }, /*#__PURE__*/_react.default.createElement(_MoreButton.default, {
    items: contextItems
  })) : null, !count && newItem ? /*#__PURE__*/_react.default.createElement(_Badge.NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined
  }) : null);
};

MenuItem.defaultProps = {
  underlined: true
};
var _default = MenuItem;
exports.default = _default;
