// @platform-ui/navigation/Menu/MoreButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _contextMenu = _interopRequireDefault(require("@platform-ui/contextMenu"));

var _MoreHorizontal = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/MoreHorizontal"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../utils");

var _MoreButtonModule = _interopRequireDefault(require("./MoreButton.module.css"));

var _NavigationContext = require("../NavigationContext");

var MoreButton = function MoreButton(props) {
  var _cn;

  var items = props.items;
  var memoizedItems = (0, _react.useMemo)(function () {
    return items.map(function (item) {
      return (0, _utils.menuItemToContextMenu)(item);
    });
  }, [items]);

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  return /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_MoreButtonModule.default.root, (_cn = {}, _cn[_MoreButtonModule.default["root_theme_" + theme]] = theme, _cn))
  }, /*#__PURE__*/_react.default.createElement(_contextMenu.default, Object.assign({
    showOnHover: true,
    highlightActive: true,
    hideOnResize: true,
    items: memoizedItems,
    itemSize: "xs"
  }, (0, _dataQa.default)(props)), /*#__PURE__*/_react.default.createElement(_MoreHorizontal.default, null)));
};

var _default = MoreButton;
exports.default = _default;
