// @platform-ui/select/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = exports.MultiselectTagged = exports.Multiselect = exports.AdaptiveSelect = void 0;

var _focusStateHOC = _interopRequireDefault(require("@platform-ui/focusStateHOC"));

var _Select = _interopRequireDefault(require("./Select"));

exports.SelectNoFocus = _Select.default;

var _NativeSelect = _interopRequireDefault(require("./NativeSelect"));

exports.NativeSelect = _NativeSelect.default;

var _Multiselect = _interopRequireDefault(require("./Multiselect/Multiselect"));

exports.MultiselectNoFocus = _Multiselect.default;

var _NativeMultiselect = _interopRequireDefault(require("./Multiselect/NativeMultiselect"));

exports.NativeMultiselect = _NativeMultiselect.default;

var _MultiselectTagged = _interopRequireDefault(require("./MultiselectTagged/MultiselectTagged"));

exports.MultiselectTaggedNoFocus = _MultiselectTagged.default;

var _NativeMultiselectTagged = _interopRequireDefault(require("./MultiselectTagged/NativeMultiselectTagged"));

exports.NativeMultiselectTagged = _NativeMultiselectTagged.default;

var _AdaptiveSelect = _interopRequireDefault(require("./AdaptiveSelect/AdaptiveSelect"));

var Select = (0, _focusStateHOC.default)(_Select.default);
var Multiselect = (0, _focusStateHOC.default)(_Multiselect.default);
exports.Multiselect = Multiselect;
var MultiselectTagged = (0, _focusStateHOC.default)(_MultiselectTagged.default);
exports.MultiselectTagged = MultiselectTagged;
var AdaptiveSelect = (0, _focusStateHOC.default)(_AdaptiveSelect.default);
exports.AdaptiveSelect = AdaptiveSelect;
var _default = Select;
exports.default = _default;
