// @platform-ui/checkable/Checkable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _focusUtils = require("@platform-ui/focusUtils");

var _excluded = ["dataQaType", "dataQaData", "focused", "focusVisible", "checked", "indeterminate"];

var useEventParams = function useEventParams(callback) {
  return function (event) {
    var _event$target = event.target,
        name = _event$target.name,
        value = _event$target.value,
        checked = _event$target.checked;
    callback(event, {
      name: name,
      value: value,
      checked: checked
    });
  };
};

var Checkable = /*#__PURE__*/(0, _react.forwardRef)(function (props, forwardedRef) {
  var handleBlur = useEventParams(props.onBlur);
  var handleFocus = useEventParams(props.onFocus);
  var handleChange = useEventParams(props.onChange);
  var propsWithRefAndEventParams = (0, _extends2.default)({}, props, {
    ref: forwardedRef,
    onBlur: handleBlur,
    onFocus: handleFocus,
    onChange: handleChange
  });
  var propsWithControlledFocus = (0, _focusUtils.useControlledFocus)(propsWithRefAndEventParams);
  var propsWithFocusVisible = (0, _focusUtils.useFocusVisible)(propsWithControlledFocus);
  var dataQaType = propsWithFocusVisible.dataQaType,
      dataQaData = propsWithFocusVisible.dataQaData,
      focused = propsWithFocusVisible.focused,
      focusVisible = propsWithFocusVisible.focusVisible,
      checked = propsWithFocusVisible.checked,
      indeterminate = propsWithFocusVisible.indeterminate,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(propsWithFocusVisible, _excluded);
  return /*#__PURE__*/_react.default.createElement("input", Object.assign({}, (0, _dataQa.default)({
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }), {
    role: propsWithFocusVisible.type,
    "aria-checked": indeterminate ? 'mixed' : checked,
    checked: checked
  }, restProps, {
    "data-focus-visible": focusVisible || undefined
  }));
});
var _default = Checkable;
exports.default = _default;
