// @platform-ui/utils/animation.js
"use strict";

exports.__esModule = true;
exports.default = exports.TIMING_FUNCS = void 0;
var TIMING_FUNCS = {
  // no easing, no acceleration
  LINEAR: function LINEAR(t) {
    return t;
  },
  // accelerating from zero velocity
  EASE_IN: function EASE_IN(t) {
    return t * t;
  },
  // decelerating to zero velocity
  EASE_OUT: function EASE_OUT(t) {
    return t * (2 - t);
  }
};
exports.TIMING_FUNCS = TIMING_FUNCS;

var Animation = /*#__PURE__*/function () {
  function Animation() {
    var _this = this;

    this.animate = function (time) {
      var timeStamp = Animation.hasPerformance ? time : Date.now(); // timeFraction от 0 до 1

      var timeFraction = (timeStamp - _this.start) / _this.options.duration;

      if (timeFraction > 1) {
        timeFraction = 1;
      } // текущее состояние анимации


      var progress = _this.options.timing(timeFraction);

      _this.options.draw(progress);

      if (timeFraction < 1) {
        _this.id = requestAnimationFrame(_this.animate);
      }
    };

    this.id = null;
  }

  var _proto = Animation.prototype;

  _proto.play = function play(options) {
    this.options = options;
    this.start = Animation.hasPerformance ? performance.now() : Date.now();
    this.id = requestAnimationFrame(this.animate);
    return this.id;
  };

  _proto.stop = function stop() {
    cancelAnimationFrame(this.id);
  };

  return Animation;
}();

exports.default = Animation;
Animation.hasPerformance = typeof performance !== 'undefined';
