// @platform-ui/pageHeader/EditableRow/Controls.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _button = _interopRequireDefault(require("@platform-ui/button"));

var _Tip = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/Tip"));

var _Close = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/Close"));

var _ControlsModule = _interopRequireDefault(require("./Controls.module.css"));

var Controls = function Controls(props) {
  var _cn;

  var onSave = props.onSave,
      onCancel = props.onCancel,
      mobile = props.mobile,
      _props$saveButtonLabe = props.saveButtonLabel,
      saveButtonLabel = _props$saveButtonLabe === void 0 ? 'Сохранить' : _props$saveButtonLabe,
      _props$cancelButtonLa = props.cancelButtonLabel,
      cancelButtonLabel = _props$cancelButtonLa === void 0 ? 'Отменить' : _props$cancelButtonLa;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_ControlsModule.default.root, (_cn = {}, _cn[_ControlsModule.default.root_mobile] = mobile, _cn))
  }, /*#__PURE__*/_react.default.createElement(_button.default, {
    size: "s",
    theme: "flat",
    onClick: onCancel,
    icon: mobile ? _Close.default : null,
    ariaLabel: mobile ? cancelButtonLabel : undefined
  }, mobile ? null : cancelButtonLabel), /*#__PURE__*/_react.default.createElement(_button.default, {
    size: "s",
    theme: mobile ? 'flat' : 'secondary',
    onClick: onSave,
    icon: mobile ? _Tip.default : null,
    ariaLabel: mobile ? saveButtonLabel : undefined
  }, mobile ? null : saveButtonLabel));
};

var _default = Controls;
exports.default = _default;
