// @platform-ui/pageHeader/EditableRow/EditableRow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _inlineEdit = _interopRequireDefault(require("@platform-ui/inlineEdit"));

var _Row = _interopRequireDefault(require("../Row"));

var _Controls = _interopRequireDefault(require("./Controls"));

var _EditableRowModule = _interopRequireDefault(require("./EditableRow.module.css"));

var EditableRow = function EditableRow(props) {
  var _cn;

  var onChange = props.onChange,
      onSave = props.onSave,
      onCancel = props.onCancel,
      right = props.right,
      mobile = props.mobile,
      saveButtonLabel = props.saveButtonLabel,
      cancelButtonLabel = props.cancelButtonLabel;

  var _useState = (0, _react.useState)(props.value),
      value = _useState[0],
      setValue = _useState[1];

  var _useState2 = (0, _react.useState)(props.value),
      newValue = _useState2[0],
      setNewValue = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      editable = _useState3[0],
      setEditable = _useState3[1];

  var handleChange = function handleChange(event, params) {
    setNewValue(params == null ? void 0 : params.value);
    onChange && onChange(event, params);
  };

  var handleSave = function handleSave(event) {
    setValue(newValue);
    onSave && onSave(event);
  };

  var handleCancel = function handleCancel(event) {
    setValue('');
    window.requestAnimationFrame(function () {
      setValue(props.value);
    });
    setNewValue(props.value);
    onCancel && onCancel(event);
  };

  var handleFocus = function handleFocus() {
    setEditable(true);
  };

  var handleBlur = function handleBlur() {
    setEditable(false);
  };

  return /*#__PURE__*/_react.default.createElement(_Row.default, {
    left: /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)(_EditableRowModule.default.editContainer, (_cn = {}, _cn[_EditableRowModule.default.editContainer_mobile] = mobile, _cn))
    }, /*#__PURE__*/_react.default.createElement(_inlineEdit.default, {
      value: value,
      onChange: handleChange,
      onFocus: handleFocus,
      onBlur: handleBlur,
      hasIcon: !editable && value === newValue,
      iconSize: "l"
    })),
    right: !editable && value === newValue ? right : /*#__PURE__*/_react.default.createElement(_Controls.default, {
      mobile: mobile,
      onSave: handleSave,
      onCancel: handleCancel,
      saveButtonLabel: saveButtonLabel,
      cancelButtonLabel: cancelButtonLabel
    }),
    mobile: mobile
  });
};

var _default = EditableRow;
exports.default = _default;
