// @platform-ui/inputTag/Droplist/Droplist.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dropdown = require("@platform-ui/dropdown");

var _TagsContext = require("../TagsContext");

var Droplist = function Droplist(_ref) {
  var children = _ref.children;
  var context = (0, _TagsContext.useTagsContextAutocomplete)();
  var size = context.size,
      opened = context.opened,
      loading = context.loading,
      noResults = context.noResults,
      options = context.options,
      editing = context.editing,
      setEditing = context.setEditing,
      focusedIndex = context.focusedIndex,
      setInputValue = context.setInputValue,
      highlightedOption = context.highlightedOption,
      setHighlightedOption = context.setHighlightedOption,
      onAdd = context.onAdd,
      onReplace = context.onReplace,
      dataQaType = context.dataQaType,
      acceptTag = context.acceptTag;
  return /*#__PURE__*/_react.default.createElement(_dropdown.DropdownList, {
    dataQaType: dataQaType + ".droplist",
    size: size,
    opened: opened,
    options: options,
    maxHeight: loading || noResults ? 'auto' : undefined,
    highlightedOption: highlightedOption || undefined,
    onOptionClick: function onOptionClick(event, params) {
      var tag = acceptTag(params.option);

      if (tag === null) {
        return;
      }

      setHighlightedOption(null);

      if (editing) {
        setEditing(false);
        onReplace(event, {
          index: focusedIndex,
          tag: tag
        });
      } else {
        onAdd(event, {
          tag: tag
        });
      }
    },
    onOptionMouseOver: function onOptionMouseOver(event, params) {
      setHighlightedOption(params.option);
    },
    onMouseLeave: function onMouseLeave() {
      setHighlightedOption(null);
    }
  }, children);
};

var _default = Droplist;
exports.default = _default;
