// @platform-ui/navigation/ContextMenu/ContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _contextMenu = _interopRequireDefault(require("@platform-ui/contextMenu"));

var _ArrowDown = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/ArrowDown"));

var _utils = require("../utils");

var _ContextMenuModule = _interopRequireDefault(require("./ContextMenu.module.css"));

var _NavigationContext = require("../NavigationContext");

function ContextMenu(_ref) {
  var _cn, _cn2;

  var items = _ref.items,
      _ref$withTitle = _ref.withTitle,
      withTitle = _ref$withTitle === void 0 ? true : _ref$withTitle;

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  var _useState = (0, _react.useState)(false),
      opened = _useState[0],
      setOpened = _useState[1];

  var selectedItem = ((items == null ? void 0 : items.filter(function (item) {
    return item.active === true;
  })[0]) || items[0]).title;
  var memoizedItems = (0, _react.useMemo)(function () {
    return items.map(function (item) {
      return (0, _utils.menuItemToContextMenu)(item);
    });
  }, [items]);

  var handleShow = function handleShow() {
    setOpened(true);
  };

  var handleClose = function handleClose() {
    setOpened(false);
  };

  return /*#__PURE__*/_react.default.createElement(_contextMenu.default, {
    showOnHover: true,
    items: memoizedItems,
    itemSize: "xs",
    hideOnResize: true,
    onShow: handleShow,
    onClose: handleClose
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_ContextMenuModule.default.root, (_cn = {}, _cn[_ContextMenuModule.default["root_theme_" + theme]] = theme, _cn[_ContextMenuModule.default.root_active] = opened, _cn))
  }, withTitle ? /*#__PURE__*/_react.default.createElement("span", {
    className: _ContextMenuModule.default.content
  }, selectedItem) : null, /*#__PURE__*/_react.default.createElement("span", {
    className: (0, _classnames.default)(_ContextMenuModule.default.arrow, (_cn2 = {}, _cn2[_ContextMenuModule.default.arrow_opened] = opened, _cn2))
  }, /*#__PURE__*/_react.default.createElement(_ArrowDown.default, null))));
}

ContextMenu.displayName = 'Navigation.ContextMenu';
var _default = ContextMenu;
exports.default = _default;
