// @platform-ui/navigation/User/Login/Login.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("../../Link"));

var _Login = _interopRequireDefault(require("../Icons/Login"));

var _LoginModule = _interopRequireDefault(require("./Login.module.css"));

var Login = function Login(props) {
  var title = props.title,
      path = props.path,
      onClick = props.onClick,
      mobile = props.mobile,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData;
  return /*#__PURE__*/_react.default.createElement(_Link.default, {
    className: _LoginModule.default.root,
    path: path,
    onClick: onClick,
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _LoginModule.default.title
  }, title), !mobile ? /*#__PURE__*/_react.default.createElement("span", {
    className: _LoginModule.default.icon
  }, /*#__PURE__*/_react.default.createElement(_Login.default, null)) : null);
};

var _default = Login;
exports.default = _default;
