// @platform-ui/label/esm/withLabel.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["dataQaType", "dataQaData", "children", "size", "disabled", "hintBlock"];
import React from 'react';
import Label from './Label';
import styles from './Label.module.css';
export const withLabel = (Component, defaults = {}) => props => {
  const {
    dataQaType = defaults.dataQaType,
    dataQaData,
    children,
    size = 'l',
    disabled,
    hintBlock
  } = props,
        inputProps = _objectWithoutPropertiesLoose(props, _excluded);

  const input = /*#__PURE__*/React.createElement(Component, Object.assign({}, inputProps, {
    size: size,
    disabled: disabled,
    dataQaType: `${dataQaType}.input`
  }));
  return /*#__PURE__*/React.createElement(Label, {
    dataQaType: dataQaType,
    dataQaData: dataQaData,
    size: size,
    input: input,
    disabled: disabled
  }, children, hintBlock ? /*#__PURE__*/React.createElement("span", {
    className: styles.hintBlock
  }, hintBlock) : null);
};
