// @platform-ui/navigation/User/UserDesktop/UserDesktopRoot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _popover = _interopRequireDefault(require("@platform-ui/popover"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _Menu = _interopRequireDefault(require("../Icons/Menu"));

var _UserDesktopRootModule = _interopRequireDefault(require("./UserDesktopRoot.module.css"));

var _NavigationContext = require("../../NavigationContext");

var UserDesktopRoot = function UserDesktopRoot(props) {
  var _cn;

  var menuComponent = props.menuComponent,
      profileComponent = props.children;

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  var _useState = (0, _react.useState)(false),
      popoverVisible = _useState[0],
      setPopoverVisible = _useState[1];

  var popoverBlock = /*#__PURE__*/_react.default.createElement("div", {
    className: _UserDesktopRootModule.default.popoverBlock
  }, /*#__PURE__*/_react.default.createElement("div", Object.assign({
    className: _UserDesktopRootModule.default.content
  }, (0, _dataQa.default)(props)), /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, typeof profileComponent === 'function' ? profileComponent({
    closeProfilePopover: function closeProfilePopover() {
      return setPopoverVisible(false);
    }
  }) : profileComponent)));

  return /*#__PURE__*/_react.default.createElement("div", {
    className: _UserDesktopRootModule.default.root
  }, menuComponent ? /*#__PURE__*/_react.default.createElement("div", {
    className: _UserDesktopRootModule.default.menuComponent
  }, menuComponent, /*#__PURE__*/_react.default.createElement("div", {
    className: _UserDesktopRootModule.default.divider
  })) : null, /*#__PURE__*/_react.default.createElement("div", {
    className: _UserDesktopRootModule.default.popover
  }, /*#__PURE__*/_react.default.createElement(_popover.default, {
    toggleOnClick: true,
    isActive: popoverVisible,
    direction: "right",
    popoverBlock: popoverBlock,
    closeOnPopoverClick: false,
    onShow: function onShow() {
      return setPopoverVisible(true);
    },
    onClose: function onClose() {
      return setPopoverVisible(false);
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_UserDesktopRootModule.default.icon, (_cn = {}, _cn[_UserDesktopRootModule.default.icon_active] = popoverVisible, _cn[_UserDesktopRootModule.default["icon_theme_" + theme]] = theme, _cn))
  }, /*#__PURE__*/_react.default.createElement(_Menu.default, null)))));
};

var _default = UserDesktopRoot;
exports.default = _default;
