// @platform-ui/navigation/User/UserDesktop/UserDesktop.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _Logout = _interopRequireDefault(require("../Logout"));

var _Profile = _interopRequireDefault(require("../Profile"));

var _Companies = _interopRequireDefault(require("../Companies"));

var _index = require("../../index");

var _UserDesktopRoot = _interopRequireDefault(require("./UserDesktopRoot"));

var UserDesktop = function UserDesktop(props) {
  var _user$companies;

  var user = props.user,
      logoutLink = props.logoutLink,
      onClick = props.onClick,
      menuComponent = props.menuComponent;
  return /*#__PURE__*/_react.default.createElement(_UserDesktopRoot.default, Object.assign({
    menuComponent: menuComponent
  }, (0, _dataQa.default)(props)), /*#__PURE__*/_react.default.createElement(_Companies.default, {
    companies: user.companies,
    onClick: onClick,
    profileLink: user.profileLink
  }), (_user$companies = user.companies) != null && _user$companies.length ? /*#__PURE__*/_react.default.createElement(_index.Divider, null) : null, /*#__PURE__*/_react.default.createElement(_Profile.default, {
    avatarUrl: user.avatarUrl,
    fullName: user.fullName,
    premium: user.premium,
    links: user.profileSections,
    onClick: onClick,
    badgeSubscriptions: user.badgeSubscriptions
  }), /*#__PURE__*/_react.default.createElement(_Logout.default, Object.assign({
    onClick: onClick
  }, logoutLink)));
};

var _default = UserDesktop;
exports.default = _default;
