// @platform-ui/navigation/esm/AccordionMenu/AccordionMenuLink.js
import React from 'react';
import cn from 'classnames';
import Link from '../Link';
import Badge, { NewSection } from '../Badge';
import styles from './AccordionMenuLink.module.css';

const AccordionMenuLink = props => {
  const {
    title,
    active,
    path,
    icon,
    count,
    onClick,
    dataQaType,
    dataQaData,
    newItem
  } = props;
  return /*#__PURE__*/React.createElement("span", {
    className: styles.root,
    "data-qa-type": "uikit/navigation.accordionMenuLink"
  }, /*#__PURE__*/React.createElement(Link, {
    className: cn(styles.link, {
      [styles.link_active]: active
    }),
    "aria-current": active ? 'page' : undefined,
    path: path,
    onClick: onClick,
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }, /*#__PURE__*/React.createElement("span", {
    className: styles.title
  }, title), icon, count ? /*#__PURE__*/React.createElement(Badge, {
    count: count
  }) : null, !count && newItem ? /*#__PURE__*/React.createElement("span", {
    className: styles.newItem
  }, /*#__PURE__*/React.createElement(NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined,
    sup: false
  })) : null));
};

export default AccordionMenuLink;
