// @platform-ui/navigation/esm/AccordionMenu/AccordionMenuItem.js
import React from 'react';
import cn from 'classnames';
import dataQa from '@platform-ui/dataQa';
import CollapseBlock from '@platform-ui/collapseBlock';
import ArrowDown from '@platform-ui/iconsPack/interface/24/ArrowDown';
import Divider from '../Divider';
import AccordionMenuLink from './AccordionMenuLink';
import styles from './AccordionMenuItem.module.css';
import { NewSection } from '../Badge';

const AccordionMenuItem = props => {
  const {
    title,
    items: menuItems,
    expanded,
    alwaysExpanded,
    onExpand,
    onClick,
    lastChild,
    active,
    dataQaData,
    dataQaType = 'uikit/navigation.accordionMenuItem',
    dropdownItems,
    newItem
  } = props;
  const items = dropdownItems || menuItems;

  if (!(items != null && items.length)) {
    return /*#__PURE__*/React.createElement("div", Object.assign({}, dataQa({
      dataQaType,
      dataQaData
    }), {
      className: styles.root
    }), /*#__PURE__*/React.createElement("div", {
      className: styles.row
    }, /*#__PURE__*/React.createElement(AccordionMenuLink, Object.assign({}, props, {
      active: active,
      onClick: onClick
    }))), lastChild ? null : /*#__PURE__*/React.createElement(Divider, null));
  }

  return /*#__PURE__*/React.createElement("div", Object.assign({}, dataQa({
    dataQaType: cn(dataQaType, {
      [`${dataQaType}_expanded`]: expanded
    }),
    dataQaData
  }), {
    className: cn(styles.root, {
      [styles.root_expanded]: expanded,
      [styles.root_active]: active
    })
  }), /*#__PURE__*/React.createElement("div", {
    className: styles.row,
    onClick: onExpand
  }, /*#__PURE__*/React.createElement("span", Object.assign({}, dataQa({
    dataQaType: `${dataQaType}.title`
  }), {
    className: styles.title
  }), title), newItem ? /*#__PURE__*/React.createElement("span", {
    className: styles.newItem
  }, /*#__PURE__*/React.createElement(NewSection, {
    content: typeof newItem === 'string' ? newItem : undefined,
    sup: false
  })) : null, !alwaysExpanded ? /*#__PURE__*/React.createElement("span", {
    className: styles.iconWrapper
  }, /*#__PURE__*/React.createElement("span", Object.assign({}, dataQa({
    dataQaType: `${dataQaType}.icon`
  }), {
    className: styles.icon
  }), /*#__PURE__*/React.createElement(ArrowDown, null))) : null), /*#__PURE__*/React.createElement(CollapseBlock, {
    collapsed: !alwaysExpanded && !expanded
  }, /*#__PURE__*/React.createElement("div", {
    className: styles.items
  }, items.map((itemProps, index) => {
    return /*#__PURE__*/React.createElement(AccordionMenuLink // eslint-disable-next-line react/no-array-index-key
    , Object.assign({
      key: index,
      onClick: onClick,
      dataQaType: itemProps.dataQaType || `${dataQaType}Second`
    }, itemProps));
  }))), lastChild ? null : /*#__PURE__*/React.createElement(Divider, null));
};

export default AccordionMenuItem;
