// @platform-ui/navigation/esm/AccordionMenu/AccordionMenu.js
import React, { useState, useMemo, useEffect } from 'react';
import AccordionMenuItem from './AccordionMenuItem';
import styles from './AccordionMenu.module.css';

const AccordionMenu = props => {
  const {
    items,
    onClick,
    dataQaType,
    dataQaData,
    id
  } = props;
  const activeIndex = useMemo(() => items.findIndex(item => Boolean(item.active)), [items]);
  const [expanded, setExpanded] = useState();

  const handleExpand = index => event => {
    event.stopPropagation();
    setExpanded(index === expanded ? undefined : index);
  };

  useEffect(() => {
    setExpanded(activeIndex);
  }, [activeIndex]);
  return /*#__PURE__*/React.createElement("div", {
    id: id,
    className: styles.root,
    "data-qa-type": dataQaType || 'uikit/navigation.accordionMenu',
    "data-qa-data": dataQaData
  }, items.map((itemProps, index) => {
    return /*#__PURE__*/React.createElement(AccordionMenuItem // eslint-disable-next-line react/no-array-index-key
    , Object.assign({
      key: index,
      onClick: onClick,
      expanded: index === expanded,
      onExpand: handleExpand(index),
      lastChild: index === items.length - 1
    }, itemProps));
  }));
};

export default AccordionMenu;
