// @platform-ui/navigation/esm/MobilePopup/MobilePopup.js
import React, { useState, useRef, useEffect } from 'react';
import cn from 'classnames';
import { disableBodyScroll, clearAllBodyScrollLocks } from 'body-scroll-lock';
import dataQa from '@platform-ui/dataQa';
import Draggable from '@platform-ui/draggable';
import Portal from '@platform-ui/portal';
import NavClose from '@platform-ui/iconsPack/interface/24/NavClose';
import styles from './MobilePopup.module.css';
import Row from '../Row';
import Background from '../Background';
import Container from '../Container';
const DELTA_HIDE = 50; // percents

const ALIGN_LEFT = 'left';
const ALIGN_RIGHT = 'right';

const MobilePopup = props => {
  const {
    children,
    active,
    logo,
    align,
    onClose
  } = props;
  const [dragDelta, setDragDelta] = useState(0);
  const [dragClose, setDragClose] = useState(false);
  const scrollContainerEl = useRef(null);

  const handleClose = () => {
    onClose();
  };

  const handleDrag = event => {
    const {
      delta: {
        page: {
          x: delta
        }
      },
      rect: {
        width
      }
    } = event;

    if (align === ALIGN_LEFT && delta >= 0) {
      return;
    }

    if (align === ALIGN_RIGHT && delta <= 0) {
      return;
    }

    setDragDelta(delta);
    setDragClose(Math.abs(delta) >= width / 100 * DELTA_HIDE);
  };

  const handleDragEnd = () => {
    if (dragClose) {
      handleClose();
    }

    setDragDelta(0);
    setDragClose(false);
  };

  useEffect(() => {
    if (active && scrollContainerEl.current) {
      disableBodyScroll(scrollContainerEl.current);
    }

    return () => {
      clearAllBodyScrollLocks();
    };
  }, [active]);
  return /*#__PURE__*/React.createElement(Portal, {
    toBody: true
  }, /*#__PURE__*/React.createElement("nav", Object.assign({
    className: cn(styles.root, {
      [styles.root_active]: active,
      [styles[`root_align-${align}`]]: true
    })
  }, dataQa(props)), /*#__PURE__*/React.createElement("div", {
    className: styles.overlay
  }), /*#__PURE__*/React.createElement("div", {
    className: styles.wrapper,
    style: {
      left: dragDelta
    }
  }, /*#__PURE__*/React.createElement(Draggable, {
    onDrag: handleDrag,
    onDragEnd: handleDragEnd,
    onDragStart: handleDrag,
    allowDefaultAction: true,
    touchable: true,
    style: {
      height: '100%'
    },
    onClick: handleClose
  }, /*#__PURE__*/React.createElement(Background, {
    theme: "gray"
  }, /*#__PURE__*/React.createElement(Container, {
    mobile: true
  }, /*#__PURE__*/React.createElement(Row, {
    mobile: true,
    left: logo,
    right: /*#__PURE__*/React.createElement(NavClose, {
      dataQaType: "uikit/navigation.closeMobilePopup"
    })
  }))), /*#__PURE__*/React.createElement("div", {
    className: styles.scrollContainer,
    ref: scrollContainerEl
  }, children)))));
};

MobilePopup.defaultProps = {
  align: 'left',
  dataQaType: 'uikit/navigation.MobilePopup'
};
export default MobilePopup;
