// @platform-ui/island/IslandPager/ControllableIslandPager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _carousel = _interopRequireDefault(require("@platform-ui/carousel"));

var _Island = _interopRequireDefault(require("../Island"));

var _types = require("../types");

var _IslandPagerModule = _interopRequireDefault(require("./IslandPager.module.css"));

var _IslandPagerButton = _interopRequireDefault(require("./IslandPagerButton"));

var _excluded = ["onChange", "activeIndex", "children", "autoHeight", "formatPaginationText"];

var ControllableIslandPager = function ControllableIslandPager(props) {
  var onChange = props.onChange,
      activeIndex = props.activeIndex,
      children = props.children,
      _props$autoHeight = props.autoHeight,
      autoHeight = _props$autoHeight === void 0 ? false : _props$autoHeight,
      _props$formatPaginati = props.formatPaginationText,
      formatPaginationText = _props$formatPaginati === void 0 ? function (currentPage, totalPages) {
    return currentPage + " \u0438\u0437 " + totalPages;
  } : _props$formatPaginati,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

  var childrenCount = _react.Children.count(children);

  var paginationText = formatPaginationText(activeIndex + 1, childrenCount);

  var handleUpdateActiveIndexByCarousel = function handleUpdateActiveIndexByCarousel(index) {
    onChange(index);
  };

  var handlePrevClick = function handlePrevClick() {
    onChange(activeIndex - 1);
  };

  var handleNextClick = function handleNextClick() {
    onChange(activeIndex + 1);
  };

  return /*#__PURE__*/_react.default.createElement(_Island.default, restProps, /*#__PURE__*/_react.default.createElement("header", {
    className: _IslandPagerModule.default.header
  }, /*#__PURE__*/_react.default.createElement(_IslandPagerButton.default, {
    direction: "left",
    disabled: activeIndex === 0,
    onClick: handlePrevClick
  }), /*#__PURE__*/_react.default.createElement(_IslandPagerButton.default, {
    direction: "right",
    disabled: activeIndex === childrenCount - 1,
    onClick: handleNextClick
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: _IslandPagerModule.default.counter
  }, paginationText)), /*#__PURE__*/_react.default.createElement(_carousel.default, {
    activeIndex: activeIndex,
    setActive: handleUpdateActiveIndexByCarousel,
    autoHeight: autoHeight
  }, children));
};

ControllableIslandPager.defaultProps = _types.DEFAULT_PROPS;
var _default = ControllableIslandPager;
exports.default = _default;
