// @platform-ui/island/IslandPager/IslandPagerButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _clickable = _interopRequireDefault(require("@platform-ui/clickable"));

var _ArrowDown = _interopRequireDefault(require("@platform-ui/iconsPack/interface/16/ArrowDown"));

var _IslandPagerModule = _interopRequireDefault(require("./IslandPager.module.css"));

var IslandPagerButton = function IslandPagerButton(props) {
  var _cn;

  var direction = props.direction,
      disabled = props.disabled,
      onClick = props.onClick;
  return /*#__PURE__*/_react.default.createElement(_clickable.default, {
    className: (0, _classnames.default)(_IslandPagerModule.default.button, (_cn = {}, _cn[_IslandPagerModule.default.button_disabled] = disabled, _cn)),
    innerClassName: _IslandPagerModule.default.buttonInner,
    disabled: disabled,
    onClick: onClick
  }, /*#__PURE__*/_react.default.createElement(_ArrowDown.default, {
    rotate: direction === 'left' ? 90 : 270
  }));
};

var _default = IslandPagerButton;
exports.default = _default;
