// @platform-ui/popup/esm/templates/Error/Error.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["title", "description", "approve", "decline", "onClose", "size"];
import React, { useCallback } from 'react';
import { useMedia } from 'react-use';
import cn from 'classnames';
import Button, { ButtonLink } from '@platform-ui/button';
import AlertIcon from '@platform-ui/iconsPack/markers/ui/Alert';
import mediaBreakpoints from '@platform-ui/constants/mediaBreakpoints';
import Popup from '../../Popup';
import styles from './Error.module.css';

function Error(_ref) {
  let {
    title,
    description,
    approve,
    decline,
    onClose,
    size = 'm'
  } = _ref,
      otherProps = _objectWithoutPropertiesLoose(_ref, _excluded);

  const {
    tabletS
  } = mediaBreakpoints;
  const mediaTablet = useMedia(tabletS, true);
  const handleClose = useCallback((event, params) => {
    onClose && onClose(event, {
      value: params && typeof params.value !== 'undefined' ? params.value : undefined
    });
  }, [onClose]);
  return /*#__PURE__*/React.createElement(Popup, Object.assign({}, otherProps, {
    size: size,
    mobileSize: "fullscreen",
    onClose: handleClose
  }), /*#__PURE__*/React.createElement("div", {
    style: {
      display: 'flex',
      flexDirection: 'column',
      minHeight: mediaTablet ? 'auto' : 'calc(100vh - 64px)'
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: cn(styles.icon, {
      [styles[`icon_desktop-${size}`]]: mediaTablet,
      [styles.icon_mobile]: !mediaTablet
    })
  }, /*#__PURE__*/React.createElement(AlertIcon, {
    theme: "red",
    withContainer: true,
    iconSize: !mediaTablet || size === 'm' ? 96 : 64
  })), /*#__PURE__*/React.createElement("h2", {
    className: cn(styles.title, {
      [styles.title_desktop]: mediaTablet,
      [styles.title_mobile]: !mediaTablet
    })
  }, title), /*#__PURE__*/React.createElement("div", {
    className: styles.content
  }, description), /*#__PURE__*/React.createElement("div", {
    className: styles.buttons
  }, /*#__PURE__*/React.createElement(Button, {
    size: size === 'm' && mediaTablet ? 'l' : 'm',
    wide: !mediaTablet,
    onClick: useCallback(event => {
      handleClose(event, {
        value: true
      });
    }, [handleClose])
  }, approve), decline && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("span", {
    className: styles.separator
  }), /*#__PURE__*/React.createElement(ButtonLink, {
    size: size === 'm' && mediaTablet ? 'l' : 'm',
    wide: !mediaTablet,
    onClick: event => {
      handleClose(event, {
        value: false
      });
    }
  }, decline)))));
}

export default Error;
