// @platform-ui/notification/NotificationFullScreen/NotificationFullScreen.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _iconsPack = require("@platform-ui/iconsPack");

var _link = _interopRequireDefault(require("@platform-ui/link"));

var _NotificationFullScreenModule = _interopRequireDefault(require("./NotificationFullScreen.module.css"));

var iconInfo = /*#__PURE__*/_react.default.createElement(_iconsPack.Int24Attention, null);

var iconSuccess = /*#__PURE__*/_react.default.createElement(_iconsPack.Int24Success, null);

var NotificationFullScreen = function NotificationFullScreen(props) {
  var type = props.type,
      withIcon = props.withIcon,
      icon = props.icon,
      children = props.children,
      linkText = props.linkText,
      linkHref = props.linkHref,
      linkPseudo = props.linkPseudo;
  return /*#__PURE__*/_react.default.createElement("div", Object.assign({}, (0, _dataQa.default)(props), {
    className: _NotificationFullScreenModule.default.root + " " + _NotificationFullScreenModule.default["root_" + type]
  }), withIcon && /*#__PURE__*/_react.default.createElement("div", {
    className: _NotificationFullScreenModule.default.icon
  }, !icon && (type === 'success' && iconSuccess || iconInfo), icon), /*#__PURE__*/_react.default.createElement("div", {
    className: _NotificationFullScreenModule.default.content
  }, children, linkText && /*#__PURE__*/_react.default.createElement("div", {
    className: _NotificationFullScreenModule.default.link
  }, /*#__PURE__*/_react.default.createElement(_link.default, {
    href: linkHref,
    pseudo: linkPseudo
  }, linkText))));
}; // TODO: убрать в мажорном обновлении


NotificationFullScreen.defaultProps = {
  dataQaType: 'uikit/notificationFullScreen',
  type: 'info',
  withIcon: true
};
var _default = NotificationFullScreen;
exports.default = _default;
