// @platform-ui/inputSearchTagged/InputSearchTagged.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _InputTagAutocomplete = _interopRequireDefault(require("@platform-ui/inputTag/InputTagAutocomplete"));

var _Search = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/Search"));

var _InputSearchTaggedModule = _interopRequireDefault(require("./InputSearchTagged.module.css"));

var _excluded = ["onSearchIconClick"];

var InputSearchTagged = function InputSearchTagged(_ref) {
  var onSearchIconClick = _ref.onSearchIconClick,
      props = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  return /*#__PURE__*/_react.default.createElement(_InputTagAutocomplete.default, Object.assign({
    leftContent: /*#__PURE__*/_react.default.createElement("span", {
      className: _InputSearchTaggedModule.default.icon,
      onClick: onSearchIconClick
    }, /*#__PURE__*/_react.default.createElement(_Search.default, null))
  }, props));
};

InputSearchTagged.defaultProps = {
  cleanable: true,
  scrollable: true
};
var _default = InputSearchTagged;
exports.default = _default;
