// @platform-ui/block/BlockHint.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.BlockHint = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _tooltip = require("@platform-ui/tooltip");

var _BlockHintWrapper = require("./BlockHintWrapper");

var _excluded = ["dataQaType", "children"];

var BlockHint = /*#__PURE__*/_react.default.memo(function (props) {
  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/block.blockHint' : _props$dataQaType,
      children = props.children,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  return /*#__PURE__*/_react.default.createElement(_BlockHintWrapper.BlockHintWrapper, {
    dataQaType: dataQaType + ".wrapper"
  }, /*#__PURE__*/_react.default.createElement(_tooltip.Hint, Object.assign({
    popoverContent: children,
    preventDefaultClick: true,
    theme: "dark",
    direction: "bottom",
    align: "end",
    additionalShift: 12
  }, otherProps, {
    dataQaType: dataQaType + ".content"
  })));
});

exports.BlockHint = BlockHint;
