// @platform-ui/block/Block.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var corners = _interopRequireWildcard(require("@platform-ui/flatCorners"));

var _BlockSection = require("./BlockSection");

var _PrimaryText = require("./PrimaryText");

var _SecondaryText = require("./SecondaryText");

var _BlockHintWrapper = require("./BlockHintWrapper");

var _BlockModule = _interopRequireDefault(require("./Block.module.css"));

var Block = function Block(props) {
  var _cn;

  var _props$dataQaType = props.dataQaType,
      dataQaType = _props$dataQaType === void 0 ? 'uikit/block' : _props$dataQaType,
      disabled = props.disabled,
      input = props.input,
      icon = props.icon,
      title = props.title,
      label = props.label,
      description = props.description,
      hintBlock = props.hintBlock,
      _props$size = props.size,
      size = _props$size === void 0 ? 'l' : _props$size,
      vertical = props.vertical,
      reverse = props.reverse,
      inGroup = props.inGroup,
      separated = props.separated,
      _props$flatCorners = props.flatCorners,
      flatCorners = _props$flatCorners === void 0 ? '' : _props$flatCorners,
      _props$borderType = props.borderType,
      borderType = _props$borderType === void 0 ? 'around' : _props$borderType,
      _props$alignSections = props.alignSections,
      alignSections = _props$alignSections === void 0 ? 'smart' : _props$alignSections;

  var fieldChild = /*#__PURE__*/_react.default.createElement(_BlockSection.BlockSection, null, input);

  var children = props.children;

  if (!children) {
    var large = size === 'l' || size === 'xl';

    var hintChild = hintBlock && /*#__PURE__*/_react.default.createElement(_BlockHintWrapper.BlockHintWrapper, {
      key: "hint"
    }, hintBlock);

    var iconChild = icon && /*#__PURE__*/_react.default.createElement(_BlockSection.BlockSection, {
      centered: !title && !description,
      key: "icon"
    }, icon);

    var mainChild = /*#__PURE__*/_react.default.createElement(_BlockSection.BlockSection, {
      centered: !title && !description,
      stretch: true,
      key: "main"
    }, title && large && /*#__PURE__*/_react.default.createElement(_SecondaryText.SecondaryText, {
      role: "heading",
      "aria-level": 3,
      dataQaType: dataQaType + ".title"
    }, title), /*#__PURE__*/_react.default.createElement(_PrimaryText.PrimaryText, {
      role: "heading",
      "aria-level": 4,
      dataQaType: dataQaType + ".label"
    }, label, !reverse && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "\xA0", hintChild)), description && large && /*#__PURE__*/_react.default.createElement(_SecondaryText.SecondaryText, {
      dataQaType: dataQaType + ".description"
    }, description));

    children = reverse ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, mainChild, hintChild && /*#__PURE__*/_react.default.createElement(_BlockSection.BlockSection, null, hintChild), iconChild) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, iconChild, mainChild);
  }

  var groupMod;

  if (inGroup && borderType !== 'under') {
    groupMod = vertical ? 'vertical' : 'horizontal';
  }

  return /*#__PURE__*/_react.default.createElement("label", {
    "data-qa-type": dataQaType,
    className: (0, _classnames.default)((_cn = {}, _cn[_BlockModule.default.wrapper] = true, _cn[_BlockModule.default.wrapper_disabled] = disabled, _cn[_BlockModule.default.wrapper_reverse] = reverse, _cn[_BlockModule.default.wrapper_separated] = separated, _cn[_BlockModule.default["wrapper_size_" + size]] = size, _cn[_BlockModule.default["wrapper_inGroup_" + groupMod]] = groupMod, _cn[_BlockModule.default["wrapper_borderType_" + borderType]] = borderType, _cn[_BlockModule.default["wrapper_alignSections_" + alignSections]] = alignSections, _cn)),
    style: corners.style(flatCorners)
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: _BlockModule.default.focusWrapper
  }, reverse ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, fieldChild, children) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children, fieldChild)));
};

var _default = Block;
exports.default = _default;
