// @platform-ui/block/withBlock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

exports.__esModule = true;
exports.withBlock = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _Block = _interopRequireDefault(require("./Block"));

var _excluded = ["dataQaType", "disabled", "icon", "title", "label", "description", "hintBlock", "size", "vertical", "reverse", "inGroup", "separated", "flatCorners", "borderType", "alignSections", "children"];
var INPUT_SIZE_MAP = {
  s: 's',
  m: 's',
  l: 'l',
  xl: 'l'
};

var withBlock = function withBlock(Component, defaults) {
  if (defaults === void 0) {
    defaults = {};
  }

  return function (props) {
    var _props$dataQaType = props.dataQaType,
        dataQaType = _props$dataQaType === void 0 ? defaults.dataQaType : _props$dataQaType,
        disabled = props.disabled,
        icon = props.icon,
        title = props.title,
        label = props.label,
        description = props.description,
        hintBlock = props.hintBlock,
        _props$size = props.size,
        size = _props$size === void 0 ? 'l' : _props$size,
        vertical = props.vertical,
        reverse = props.reverse,
        inGroup = props.inGroup,
        separated = props.separated,
        flatCorners = props.flatCorners,
        borderType = props.borderType,
        alignSections = props.alignSections,
        children = props.children,
        inputProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);

    var input = /*#__PURE__*/_react.default.createElement(Component, Object.assign({}, inputProps, {
      disabled: disabled,
      size: INPUT_SIZE_MAP[size],
      dataQaType: dataQaType + ".input"
    }));

    return /*#__PURE__*/_react.default.createElement(_Block.default, {
      dataQaType: dataQaType,
      size: size,
      disabled: disabled,
      input: input,
      icon: icon,
      title: title,
      label: label,
      description: description,
      hintBlock: hintBlock,
      vertical: vertical,
      reverse: reverse,
      inGroup: inGroup,
      separated: separated,
      flatCorners: flatCorners,
      borderType: borderType,
      alignSections: alignSections
    }, children);
  };
};

exports.withBlock = withBlock;
