// @platform-ui/navigation/AccordionPopup/AccordionPopup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _NavMenu = _interopRequireDefault(require("@platform-ui/iconsPack/interface/24/NavMenu"));

var _generateId = _interopRequireDefault(require("@platform-ui/generateId"));

var _MobilePopup = _interopRequireDefault(require("../MobilePopup"));

var _AccordionMenu = _interopRequireDefault(require("../AccordionMenu"));

var _NavigationContext = require("../NavigationContext");

var _AccordionPopupModule = _interopRequireDefault(require("./AccordionPopup.module.css"));

var AccordionPopup = function AccordionPopup(props) {
  var _cn;

  var logo = props.logo,
      items = props.items,
      onClick = props.onClick,
      onShow = props.onShow,
      onHide = props.onHide,
      propsId = props.id,
      _props$buttonAriaLabe = props.buttonAriaLabel,
      buttonAriaLabel = _props$buttonAriaLabe === void 0 ? 'Открыть меню' : _props$buttonAriaLabe;

  var _useState = (0, _react.useState)(false),
      menuVisible = _useState[0],
      setMenuVisible = _useState[1];

  var id = (0, _react.useRef)(propsId != null ? propsId : (0, _generateId.default)());

  var _useNavigationContext = (0, _NavigationContext.useNavigationContext)(),
      theme = _useNavigationContext.theme;

  return /*#__PURE__*/_react.default.createElement("div", (0, _dataQa.default)(props), /*#__PURE__*/_react.default.createElement("button", {
    "data-qa-type": props.dataQaType + ".navBurger",
    type: "button",
    className: (0, _classnames.default)((_cn = {}, _cn[_AccordionPopupModule.default.navBurger] = true, _cn[_AccordionPopupModule.default["navBurger_theme_" + theme]] = theme, _cn)),
    onClick: (0, _react.useCallback)(function () {
      setMenuVisible(true);
      onShow && onShow();
    }, [onShow]),
    "aria-label": buttonAriaLabel,
    "aria-expanded": menuVisible,
    "aria-controls": id.current
  }, /*#__PURE__*/_react.default.createElement(_NavMenu.default, null)), /*#__PURE__*/_react.default.createElement(_MobilePopup.default, {
    logo: logo,
    active: menuVisible,
    onClose: (0, _react.useCallback)(function () {
      setMenuVisible(false);
      onHide && onHide();
    }, [onHide])
  }, /*#__PURE__*/_react.default.createElement(_AccordionMenu.default, {
    id: id.current,
    items: items,
    onClick: onClick
  })));
};

AccordionPopup.defaultProps = {
  dataQaType: 'uikit/navigation.accordionPopup'
};
var _default = AccordionPopup;
exports.default = _default;
