// @platform-ui/layout/Box/Box.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _dataQa = _interopRequireDefault(require("@platform-ui/dataQa"));

var _utils = require("../utils");

var _excluded = ["width", "height", "margin", "padding", "grow", "style", "tagName", "dataQaType", "dataQaData"];
var Box = /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
  var width = props.width,
      height = props.height,
      margin = props.margin,
      padding = props.padding,
      grow = props.grow,
      style = props.style,
      _props$tagName = props.tagName,
      TagName = _props$tagName === void 0 ? 'div' : _props$tagName,
      dataQaType = props.dataQaType,
      dataQaData = props.dataQaData,
      restProps = (0, _objectWithoutPropertiesLoose2.default)(props, _excluded);
  return /*#__PURE__*/_react.default.createElement(TagName, Object.assign({
    ref: ref,
    style: (0, _extends2.default)({
      width: (0, _utils.getSpacing)(width),
      height: (0, _utils.getSpacing)(height),
      margin: (0, _utils.getSpacing)(margin),
      padding: (0, _utils.getSpacing)(padding),
      flexGrow: grow
    }, style)
  }, (0, _dataQa.default)({
    dataQaType: dataQaType,
    dataQaData: dataQaData
  }), restProps));
});
var _default = Box;
exports.default = _default;
