// @platform-ui/styleProps/margin/index.js
"use strict";

exports.__esModule = true;
exports.useMarginStyle = exports.useMarginProps = exports.pickMarginProps = exports.calcMarginStyle = void 0;

var _react = require("react");

var _utils = require("../utils");

var pickMarginProps = (0, _utils.createPropsPicker)(['margin', 'marginX', 'marginY', 'marginTop', 'marginLeft', 'marginRight', 'marginBottom']);
exports.pickMarginProps = pickMarginProps;

var calcMarginStyle = function calcMarginStyle(props) {
  var m = props.margin,
      mX = props.marginX,
      mY = props.marginY,
      mTop = props.marginTop,
      mLeft = props.marginLeft,
      mRight = props.marginRight,
      mBottom = props.marginBottom;
  var mA = (0, _utils.splitShorthandProperty)(m);
  var marginTop = (0, _utils.pickFirstDefined)(mTop, mY, mA[0]);
  var marginLeft = (0, _utils.pickFirstDefined)(mLeft, mX, mA[3], mA[1], mA[0]);
  var marginRight = (0, _utils.pickFirstDefined)(mRight, mX, mA[1], mA[0]);
  var marginBottom = (0, _utils.pickFirstDefined)(mBottom, mY, mA[2], mA[0]);
  return {
    marginTop: marginTop,
    marginLeft: marginLeft,
    marginRight: marginRight,
    marginBottom: marginBottom
  };
};

exports.calcMarginStyle = calcMarginStyle;

var useMarginStyle = function useMarginStyle(props) {
  var margin = props.margin,
      marginX = props.marginX,
      marginY = props.marginY,
      marginTop = props.marginTop,
      marginLeft = props.marginLeft,
      marginRight = props.marginRight,
      marginBottom = props.marginBottom;
  return (0, _react.useMemo)(function () {
    return calcMarginStyle({
      margin: margin,
      marginX: marginX,
      marginY: marginY,
      marginTop: marginTop,
      marginLeft: marginLeft,
      marginRight: marginRight,
      marginBottom: marginBottom
    });
  }, [margin, marginX, marginY, marginTop, marginLeft, marginRight, marginBottom]);
};

exports.useMarginStyle = useMarginStyle;
var useMarginProps = (0, _utils.createPropsHook)(pickMarginProps, useMarginStyle);
exports.useMarginProps = useMarginProps;
