// @platform-ui/popup/templates/Dialog/Dialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;

exports.__esModule = true;
exports.default = void 0;

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireWildcard(require("react"));

var _reactUse = require("react-use");

var _classnames = _interopRequireDefault(require("classnames"));

var _button = _interopRequireDefault(require("@platform-ui/button"));

var _mediaBreakpoints = _interopRequireDefault(require("@platform-ui/constants/mediaBreakpoints"));

var _Popup = _interopRequireDefault(require("../../Popup"));

var _DialogModule = _interopRequireDefault(require("./Dialog.module.css"));

var _excluded = ["title", "description", "approve", "decline", "onClose"];

function Dialog(_ref) {
  var _cn, _cn2;

  var title = _ref.title,
      description = _ref.description,
      approve = _ref.approve,
      decline = _ref.decline,
      onClose = _ref.onClose,
      otherProps = (0, _objectWithoutPropertiesLoose2.default)(_ref, _excluded);
  var tabletS = _mediaBreakpoints.default.tabletS;
  var mediaTablet = (0, _reactUse.useMedia)(tabletS, true);
  var handleClose = (0, _react.useCallback)(function (event, params) {
    onClose && onClose(event, {
      value: params && typeof params.value !== 'undefined' ? params.value : undefined
    });
  }, [onClose]);
  return /*#__PURE__*/_react.default.createElement(_Popup.default, Object.assign({}, otherProps, {
    size: "s",
    mobileSize: "s",
    onClose: handleClose
  }), /*#__PURE__*/_react.default.createElement("h2", {
    className: (0, _classnames.default)(_DialogModule.default.title, (_cn = {}, _cn[_DialogModule.default.title_desktop] = mediaTablet, _cn[_DialogModule.default.title_mobile] = !mediaTablet, _cn))
  }, title), /*#__PURE__*/_react.default.createElement("div", {
    className: _DialogModule.default.content
  }, description), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)(_DialogModule.default.buttons, (_cn2 = {}, _cn2[_DialogModule.default.buttons_desktop] = mediaTablet, _cn2[_DialogModule.default.buttons_mobile] = !mediaTablet, _cn2))
  }, decline && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_button.default, {
    size: "m",
    theme: mediaTablet ? 'flat' : 'secondary',
    wide: !mediaTablet,
    onClick: function onClick(event) {
      handleClose(event, {
        value: false
      });
    }
  }, decline), /*#__PURE__*/_react.default.createElement("span", {
    className: _DialogModule.default.separator
  })), /*#__PURE__*/_react.default.createElement(_button.default, {
    size: "m",
    wide: !mediaTablet,
    onClick: (0, _react.useCallback)(function (event) {
      handleClose(event, {
        value: true
      });
    }, [handleClose])
  }, approve)));
}

var _default = Dialog;
exports.default = _default;
